// drizzle-orm/pg-core/dialect.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgDialect = void 0;
const column_1 = require("../column");
const columns_1 = require("./columns");
const table_1 = require("./table");
const sql_1 = require("../sql");
const subquery_1 = require("../subquery");
const table_2 = require("../table");
const utils_1 = require("../utils");
const view_1 = require("../view");
const view_2 = require("./view");
class PgDialect {
    migrate(migrations, session) {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableCreate = (0, sql_1.sql) `CREATE TABLE IF NOT EXISTS "drizzle"."__drizzle_migrations" (
			id SERIAL PRIMARY KEY,
			hash text NOT NULL,
			created_at bigint
		)`;
            yield session.execute((0, sql_1.sql) `CREATE SCHEMA IF NOT EXISTS "drizzle"`);
            yield session.execute(migrationTableCreate);
            const dbMigrations = yield session.all((0, sql_1.sql) `select id, hash, created_at from "drizzle"."__drizzle_migrations" order by created_at desc limit 1`);
            const lastDbMigration = dbMigrations[0];
            yield session.transaction((tx) => __awaiter(this, void 0, void 0, function* () {
                var _a, e_1, _b, _c;
                try {
                    for (var _d = true, migrations_1 = __asyncValues(migrations), migrations_1_1; migrations_1_1 = yield migrations_1.next(), _a = migrations_1_1.done, !_a;) {
                        _c = migrations_1_1.value;
                        _d = false;
                        try {
                            const migration = _c;
                            if (!lastDbMigration
                                || parseInt(lastDbMigration.created_at, 10) < migration.folderMillis) {
                                for (const stmt of migration.sql) {
                                    yield tx.execute(sql_1.sql.raw(stmt));
                                }
                                yield tx.execute((0, sql_1.sql) `insert into "drizzle"."__drizzle_migrations" ("hash", "created_at") values(${migration.hash}, ${migration.folderMillis})`);
                            }
                        }
                        finally {
                            _d = true;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (!_d && !_a && (_b = migrations_1.return)) yield _b.call(migrations_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }));
        });
    }
    escapeName(name) {
        return `"${name}"`;
    }
    escapeParam(num) {
        return `$${num + 1}`;
    }
    escapeString(str) {
        return `'${str.replace(/'/g, "''")}'`;
    }
    buildDeleteQuery({ table, where, returning }) {
        const returningSql = returning
            ? (0, sql_1.sql) ` returning ${this.buildSelection(returning, { isSingleTable: true })}`
            : undefined;
        const whereSql = where ? (0, sql_1.sql) ` where ${where}` : undefined;
        return (0, sql_1.sql) `delete from ${table}${whereSql}${returningSql}`;
    }
    buildUpdateSet(table, set) {
        const setEntries = Object.entries(set);
        const setSize = setEntries.length;
        return sql_1.sql.fromList(setEntries
            .map(([colName, value], i) => {
            const col = table[table_2.Table.Symbol.Columns][colName];
            const res = (0, sql_1.sql) `${(0, sql_1.name)(col.name)} = ${value}`;
            if (i < setSize - 1) {
                return [res, sql_1.sql.raw(', ')];
            }
            return [res];
        })
            .flat(1));
    }
    buildUpdateQuery({ table, set, where, returning }) {
        const setSql = this.buildUpdateSet(table, set);
        const returningSql = returning
            ? (0, sql_1.sql) ` returning ${this.buildSelection(returning, { isSingleTable: true })}`
            : undefined;
        const whereSql = where ? (0, sql_1.sql) ` where ${where}` : undefined;
        return (0, sql_1.sql) `update ${table} set ${setSql}${whereSql}${returningSql}`;
    }
    /**
     * Builds selection SQL with provided fields/expressions
     *
     * Examples:
     *
     * `select <selection> from`
     *
     * `insert ... returning <selection>`
     *
     * If `isSingleTable` is true, then columns won't be prefixed with table name
     */
    buildSelection(fields, { isSingleTable = false } = {}) {
        const columnsLen = fields.length;
        const chunks = fields
            .map(({ field }, i) => {
            const chunk = [];
            if (field instanceof sql_1.SQL.Aliased && field.isSelectionField) {
                chunk.push((0, sql_1.name)(field.fieldAlias));
            }
            else if (field instanceof sql_1.SQL.Aliased || field instanceof sql_1.SQL) {
                const query = field instanceof sql_1.SQL.Aliased ? field.sql : field;
                if (isSingleTable) {
                    chunk.push(new sql_1.SQL(query.queryChunks.map((c) => {
                        if (c instanceof columns_1.PgColumn) {
                            return (0, sql_1.name)(c.name);
                        }
                        return c;
                    })));
                }
                else {
                    chunk.push(query);
                }
                if (field instanceof sql_1.SQL.Aliased) {
                    chunk.push((0, sql_1.sql) ` as ${(0, sql_1.name)(field.fieldAlias)}`);
                }
            }
            else if (field instanceof column_1.Column) {
                if (isSingleTable) {
                    chunk.push((0, sql_1.name)(field.name));
                }
                else {
                    chunk.push(field);
                }
            }
            if (i < columnsLen - 1) {
                chunk.push((0, sql_1.sql) `, `);
            }
            return chunk;
        })
            .flat(1);
        return sql_1.sql.fromList(chunks);
    }
    buildSelectQuery({ withList, fields, where, having, table, joins, orderBy, groupBy, limit, offset, lockingClauses }) {
        const fieldsList = (0, utils_1.orderSelectedFields)(fields);
        fieldsList.forEach((f) => {
            if (f.field instanceof column_1.Column
                && (0, table_2.getTableName)(f.field.table)
                    !== (table instanceof subquery_1.Subquery
                        ? table[subquery_1.SubqueryConfig].alias
                        : table instanceof view_2.PgViewBase
                            ? table[view_1.ViewBaseConfig].name
                            : table instanceof sql_1.SQL
                                ? undefined
                                : (0, table_2.getTableName)(table))
                && !((table) => joins.some(({ alias }) => alias === (0, table_2.getTableName)(table)))(f.field.table)) {
                const tableName = (0, table_2.getTableName)(f.field.table);
                throw new Error(`Your "${f.path.join('->')}" field references a column "${tableName}"."${f.field.name}", but the table "${tableName}" is not part of the query! Did you forget to join it?`);
            }
        });
        const isSingleTable = joins.length === 0;
        let withSql;
        if (withList.length) {
            const withSqlChunks = [(0, sql_1.sql) `with `];
            withList.forEach((w, i) => {
                withSqlChunks.push((0, sql_1.sql) `${(0, sql_1.name)(w[subquery_1.SubqueryConfig].alias)} as (${w[subquery_1.SubqueryConfig].sql})`);
                if (i < withList.length - 1) {
                    withSqlChunks.push((0, sql_1.sql) `, `);
                }
            });
            withSqlChunks.push((0, sql_1.sql) ` `);
            withSql = sql_1.sql.fromList(withSqlChunks);
        }
        const selection = this.buildSelection(fieldsList, { isSingleTable });
        const joinsArray = [];
        joins.forEach((joinMeta, index) => {
            if (index === 0) {
                joinsArray.push((0, sql_1.sql) ` `);
            }
            const table = joinMeta.table;
            if (table instanceof table_1.PgTable) {
                const tableName = table[table_1.PgTable.Symbol.Name];
                const tableSchema = table[table_1.PgTable.Symbol.Schema];
                const origTableName = table[table_1.PgTable.Symbol.OriginalName];
                const alias = tableName === origTableName ? undefined : joinMeta.alias;
                joinsArray.push((0, sql_1.sql) `${sql_1.sql.raw(joinMeta.joinType)} join ${tableSchema ? (0, sql_1.sql) `${(0, sql_1.name)(tableSchema)}.` : undefined}${(0, sql_1.name)(origTableName)}${alias && (0, sql_1.sql) ` ${(0, sql_1.name)(alias)}`} on ${joinMeta.on}`);
            }
            else {
                joinsArray.push((0, sql_1.sql) `${sql_1.sql.raw(joinMeta.joinType)} join ${table} on ${joinMeta.on}`);
            }
            if (index < joins.length - 1) {
                joinsArray.push((0, sql_1.sql) ` `);
            }
        });
        const joinsSql = sql_1.sql.fromList(joinsArray);
        const whereSql = where ? (0, sql_1.sql) ` where ${where}` : undefined;
        const havingSql = having ? (0, sql_1.sql) ` having ${having}` : undefined;
        const orderByList = [];
        orderBy.forEach((orderByValue, index) => {
            orderByList.push(orderByValue);
            if (index < orderBy.length - 1) {
                orderByList.push((0, sql_1.sql) `, `);
            }
        });
        const orderBySql = orderByList.length > 0 ? (0, sql_1.sql) ` order by ${sql_1.sql.fromList(orderByList)}` : undefined;
        const groupByList = [];
        groupBy.forEach((groupByValue, index) => {
            groupByList.push(groupByValue);
            if (index < groupBy.length - 1) {
                groupByList.push((0, sql_1.sql) `, `);
            }
        });
        const groupBySql = groupByList.length > 0 ? (0, sql_1.sql) ` group by ${sql_1.sql.fromList(groupByList)}` : undefined;
        const limitSql = limit ? (0, sql_1.sql) ` limit ${limit}` : undefined;
        const offsetSql = offset ? (0, sql_1.sql) ` offset ${offset}` : undefined;
        let lockingClausesSql = sql_1.sql.empty();
        lockingClauses.forEach(({ strength, config }) => {
            let clauseSql = (0, sql_1.sql) ` for ${sql_1.sql.raw(strength)}`;
            if (config.of) {
                clauseSql.append((0, sql_1.sql) ` of ${config.of}`);
            }
            if (config.noWait) {
                clauseSql.append((0, sql_1.sql) ` no wait`);
            }
            else if (config.skipLocked) {
                clauseSql.append((0, sql_1.sql) ` skip locked`);
            }
            lockingClausesSql.append(clauseSql);
        });
        return (0, sql_1.sql) `${withSql}select ${selection} from ${table}${joinsSql}${whereSql}${groupBySql}${havingSql}${orderBySql}${limitSql}${offsetSql}${lockingClausesSql}`;
    }
    buildInsertQuery({ table, values, onConflict, returning }) {
        const isSingleValue = values.length === 1;
        const valuesSqlList = [];
        const columns = table[table_2.Table.Symbol.Columns];
        const colEntries = isSingleValue
            ? Object.keys(values[0]).map((fieldName) => [fieldName, columns[fieldName]])
            : Object.entries(columns);
        const insertOrder = colEntries.map(([, column]) => (0, sql_1.name)(column.name));
        values.forEach((value, valueIndex) => {
            const valueList = [];
            colEntries.forEach(([fieldName]) => {
                const colValue = value[fieldName];
                if (typeof colValue === 'undefined') {
                    valueList.push((0, sql_1.sql) `default`);
                }
                else {
                    valueList.push(colValue);
                }
            });
            valuesSqlList.push(valueList);
            if (valueIndex < values.length - 1) {
                valuesSqlList.push((0, sql_1.sql) `, `);
            }
        });
        const valuesSql = sql_1.sql.fromList(valuesSqlList);
        const returningSql = returning
            ? (0, sql_1.sql) ` returning ${this.buildSelection(returning, { isSingleTable: true })}`
            : undefined;
        const onConflictSql = onConflict ? (0, sql_1.sql) ` on conflict ${onConflict}` : undefined;
        return (0, sql_1.sql) `insert into ${table} ${insertOrder} values ${valuesSql}${onConflictSql}${returningSql}`;
    }
    buildRefreshMaterializedViewQuery({ view, concurrently, withNoData }) {
        const concurrentlySql = concurrently ? (0, sql_1.sql) ` concurrently` : undefined;
        const withNoDataSql = withNoData ? (0, sql_1.sql) ` with no data` : undefined;
        return (0, sql_1.sql) `refresh materialized view${concurrentlySql} ${view}${withNoDataSql}`;
    }
    prepareTyping(encoder) {
        if (encoder instanceof columns_1.PgJsonb || encoder instanceof columns_1.PgJson) {
            return 'json';
        }
        else if (encoder instanceof columns_1.PgNumeric) {
            return 'decimal';
        }
        else if (encoder instanceof columns_1.PgTime) {
            return 'time';
        }
        else if (encoder instanceof columns_1.PgTimestamp) {
            return 'timestamp';
        }
        else if (encoder instanceof columns_1.PgDate) {
            return 'date';
        }
        else if (encoder instanceof columns_1.PgUUID) {
            return 'uuid';
        }
        else {
            return 'none';
        }
    }
    sqlToQuery(sql) {
        return sql.toQuery({
            escapeName: this.escapeName,
            escapeParam: this.escapeParam,
            escapeString: this.escapeString,
            prepareTyping: this.prepareTyping,
        });
    }
}
exports.PgDialect = PgDialect;
