// drizzle-orm/pg-core/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pgSchema = exports.isPgSchema = exports.PgSchema = void 0;
const table_1 = require("./table");
const view_1 = require("./view");
class PgSchema {
    constructor(schemaName) {
        this.schemaName = schemaName;
        this.table = ((name, columns, extraConfig) => {
            return (0, table_1.pgTableWithSchema)(name, columns, extraConfig, this.schemaName);
        });
        this.view = ((name, columns) => {
            return (0, view_1.pgViewWithSchema)(name, columns, this.schemaName);
        });
        this.materializedView = ((name, columns) => {
            return (0, view_1.pgMaterializedViewWithSchema)(name, columns, this.schemaName);
        });
    }
}
exports.PgSchema = PgSchema;
/** @deprecated - use `instanceof PgSchema` */
function isPgSchema(obj) {
    return obj instanceof PgSchema;
}
exports.isPgSchema = isPgSchema;
function pgSchema(name) {
    if (name === 'public') {
        throw Error(`You can't specify 'public' as schema name. Postgres is using public schema by default. If you want to use 'public' schema, just use pgTable() instead of creating a schema`);
    }
    return new PgSchema(name);
}
exports.pgSchema = pgSchema;
