// drizzle-orm/pg-core/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgTransaction = exports.PgSession = exports.PreparedQuery = void 0;
const errors_1 = require("../errors");
const sql_1 = require("../sql");
const db_1 = require("./db");
class PreparedQuery {
}
exports.PreparedQuery = PreparedQuery;
class PgSession {
    constructor(dialect) {
        this.dialect = dialect;
    }
    execute(query) {
        return this.prepareQuery(this.dialect.sqlToQuery(query), undefined, undefined).execute();
    }
    all(query) {
        return this.prepareQuery(this.dialect.sqlToQuery(query), undefined, undefined).all();
    }
}
exports.PgSession = PgSession;
class PgTransaction extends db_1.PgDatabase {
    constructor(dialect, session, nestedIndex = 0) {
        super(dialect, session);
        this.nestedIndex = nestedIndex;
    }
    rollback() {
        throw new errors_1.TransactionRollbackError();
    }
    /** @internal */
    getTransactionConfigSQL(config) {
        const chunks = [];
        if (config.isolationLevel) {
            chunks.push(`isolation level ${config.isolationLevel}`);
        }
        if (config.accessMode) {
            chunks.push(config.accessMode);
        }
        if (typeof config.deferrable === 'boolean') {
            chunks.push(config.deferrable ? 'deferrable' : 'not deferrable');
        }
        return sql_1.sql.raw(chunks.join(' '));
    }
    setTransaction(config) {
        return this.session.execute((0, sql_1.sql) `set transaction ${this.getTransactionConfigSQL(config)}`);
    }
}
exports.PgTransaction = PgTransaction;
