// drizzle-orm/pg-core/table.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.pgTableCreator = exports.pgTable = exports.pgTableWithSchema = exports.PgTable = exports.ExtraConfigBuilder = exports.InlineForeignKeys = void 0;
const table_1 = require("../table");
/** @internal */
exports.InlineForeignKeys = Symbol('InlineForeignKeys');
/** @internal */
exports.ExtraConfigBuilder = Symbol('ExtraConfigBuilder');
class PgTable extends table_1.Table {
    constructor() {
        super(...arguments);
        /**@internal */
        this[_a] = [];
        /** @internal */
        this[_b] = undefined;
    }
}
_a = exports.InlineForeignKeys, _b = exports.ExtraConfigBuilder;
/** @internal */
PgTable.Symbol = Object.assign({}, table_1.Table.Symbol, {
    InlineForeignKeys: exports.InlineForeignKeys,
    ExtraConfigBuilder: exports.ExtraConfigBuilder,
});
exports.PgTable = PgTable;
/** @internal */
function pgTableWithSchema(name, columns, extraConfig, schema, baseName = name) {
    const rawTable = new PgTable(name, schema, baseName);
    const builtColumns = Object.fromEntries(Object.entries(columns).map(([name, colBuilder]) => {
        const column = colBuilder.build(rawTable);
        rawTable[exports.InlineForeignKeys].push(...colBuilder.buildForeignKeys(column, rawTable));
        return [name, column];
    }));
    const table = Object.assign(rawTable, builtColumns);
    table[table_1.Table.Symbol.Columns] = builtColumns;
    if (extraConfig) {
        table[PgTable.Symbol.ExtraConfigBuilder] = extraConfig;
    }
    return table;
}
exports.pgTableWithSchema = pgTableWithSchema;
const pgTable = (name, columns, extraConfig) => {
    return pgTableWithSchema(name, columns, extraConfig, undefined);
};
exports.pgTable = pgTable;
function pgTableCreator(customizeTableName) {
    return (name, columns, extraConfig) => {
        return pgTableWithSchema(customizeTableName(name), columns, extraConfig, undefined, name);
    };
}
exports.pgTableCreator = pgTableCreator;
