// drizzle-orm/pg-core/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makePgArray = exports.parsePgArray = exports.parsePgNestedArray = exports.getMaterializedViewConfig = exports.getViewConfig = exports.getTableConfig = void 0;
const table_1 = require("./table");
const table_2 = require("../table");
const view_1 = require("../view");
const checks_1 = require("./checks");
const foreign_keys_1 = require("./foreign-keys");
const indexes_1 = require("./indexes");
const primary_keys_1 = require("./primary-keys");
const view_2 = require("./view");
function getTableConfig(table) {
    const columns = Object.values(table[table_2.Table.Symbol.Columns]);
    const indexes = [];
    const checks = [];
    const primaryKeys = [];
    const foreignKeys = Object.values(table[table_1.PgTable.Symbol.InlineForeignKeys]);
    const name = table[table_2.Table.Symbol.Name];
    const schema = table[table_2.Table.Symbol.Schema];
    const extraConfigBuilder = table[table_1.PgTable.Symbol.ExtraConfigBuilder];
    if (typeof extraConfigBuilder !== 'undefined') {
        const extraConfig = extraConfigBuilder(table[table_2.Table.Symbol.Columns]);
        Object.values(extraConfig).forEach((builder) => {
            if (builder instanceof indexes_1.IndexBuilder) {
                indexes.push(builder.build(table));
            }
            else if (builder instanceof checks_1.CheckBuilder) {
                checks.push(builder.build(table));
            }
            else if (builder instanceof primary_keys_1.PrimaryKeyBuilder) {
                primaryKeys.push(builder.build(table));
            }
            else if (builder instanceof foreign_keys_1.ForeignKeyBuilder) {
                foreignKeys.push(builder.build(table));
            }
        });
    }
    return {
        columns,
        indexes,
        foreignKeys,
        checks,
        primaryKeys,
        name,
        schema,
    };
}
exports.getTableConfig = getTableConfig;
function getViewConfig(view) {
    return Object.assign(Object.assign({}, view[view_1.ViewBaseConfig]), view[view_2.PgViewConfig]);
}
exports.getViewConfig = getViewConfig;
function getMaterializedViewConfig(view) {
    return Object.assign(Object.assign({}, view[view_1.ViewBaseConfig]), view[view_2.PgMaterializedViewConfig]);
}
exports.getMaterializedViewConfig = getMaterializedViewConfig;
function parsePgArrayValue(arrayString, startFrom, inQuotes) {
    for (let i = startFrom; i < arrayString.length; i++) {
        const char = arrayString[i];
        if (char === '\\') {
            i++;
            continue;
        }
        if (char === '"') {
            return [arrayString.substring(startFrom, i).replace(/\\/g, ''), i + 1];
        }
        if (inQuotes) {
            continue;
        }
        if (char === ',' || char === '}') {
            return [arrayString.substring(startFrom, i).replace(/\\/g, ''), i];
        }
    }
    return [arrayString.substring(startFrom).replace(/\\/g, ''), arrayString.length];
}
function parsePgNestedArray(arrayString, startFrom = 0) {
    const result = [];
    let i = startFrom;
    let lastCharIsComma = false;
    while (i < arrayString.length) {
        const char = arrayString[i];
        if (char === ',') {
            if (lastCharIsComma || i === startFrom) {
                result.push('');
            }
            lastCharIsComma = true;
            i++;
            continue;
        }
        lastCharIsComma = false;
        if (char === '\\') {
            i += 2;
            continue;
        }
        if (char === '"') {
            const [value, startFrom] = parsePgArrayValue(arrayString, i + 1, true);
            result.push(value);
            i = startFrom;
            continue;
        }
        if (char === '}') {
            return [result, i + 1];
        }
        if (char === '{') {
            const [value, startFrom] = parsePgNestedArray(arrayString, i + 1);
            result.push(value);
            i = startFrom;
            continue;
        }
        const [value, newStartFrom] = parsePgArrayValue(arrayString, i, false);
        result.push(value);
        i = newStartFrom;
    }
    return [result, i];
}
exports.parsePgNestedArray = parsePgNestedArray;
function parsePgArray(arrayString) {
    const [result] = parsePgNestedArray(arrayString, 1);
    return result;
}
exports.parsePgArray = parsePgArray;
function makePgArray(array) {
    return `{${array.map((item) => {
        if (Array.isArray(item)) {
            return makePgArray(item);
        }
        if (typeof item === 'string' && item.includes(',')) {
            return `"${item.replace(/"/g, '\\"')}"`;
        }
        return `${item}`;
    }).join(',')}}`;
}
exports.makePgArray = makePgArray;
