// drizzle-orm/pg-core/view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pgMaterializedView = exports.pgView = exports.pgMaterializedViewWithSchema = exports.pgViewWithSchema = exports.PgMaterializedView = exports.PgMaterializedViewConfig = exports.PgView = exports.PgViewConfig = exports.PgViewBase = exports.ManualMaterializedViewBuilder = exports.MaterializedViewBuilder = exports.MaterializedViewBuilderCore = exports.ManualViewBuilder = exports.ViewBuilder = exports.DefaultViewBuilderCore = void 0;
const subquery_1 = require("../subquery");
const utils_1 = require("../utils");
const view_1 = require("../view");
const query_builders_1 = require("./query-builders");
const table_1 = require("./table");
class DefaultViewBuilderCore {
    constructor(name, schema) {
        this.name = name;
        this.schema = schema;
        this.config = {};
    }
    with(config) {
        this.config.with = config;
        return this;
    }
}
exports.DefaultViewBuilderCore = DefaultViewBuilderCore;
class ViewBuilder extends DefaultViewBuilderCore {
    as(qb) {
        if (typeof qb === 'function') {
            qb = qb(query_builders_1.queryBuilder);
        }
        const selectionProxy = new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        });
        const aliasedSelection = new Proxy(qb.getSelectedFields(), selectionProxy);
        return new Proxy(new PgView({
            pgConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: aliasedSelection,
                query: qb.getSQL().inlineParams(),
            },
        }), selectionProxy);
    }
}
exports.ViewBuilder = ViewBuilder;
class ManualViewBuilder extends DefaultViewBuilderCore {
    constructor(name, columns, schema) {
        super(name, schema);
        this.columns = (0, utils_1.getTableColumns)((0, table_1.pgTable)(name, columns));
    }
    existing() {
        return new Proxy(new PgView({
            pgConfig: undefined,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: undefined,
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
    as(query) {
        return new Proxy(new PgView({
            pgConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: query.inlineParams(),
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
}
exports.ManualViewBuilder = ManualViewBuilder;
class MaterializedViewBuilderCore {
    constructor(name, schema) {
        this.name = name;
        this.schema = schema;
        this.config = {};
    }
    using(using) {
        this.config.using = using;
        return this;
    }
    with(config) {
        this.config.with = config;
        return this;
    }
    tablespace(tablespace) {
        this.config.tablespace = tablespace;
        return this;
    }
    withNoData() {
        this.config.withNoData = true;
        return this;
    }
}
exports.MaterializedViewBuilderCore = MaterializedViewBuilderCore;
class MaterializedViewBuilder extends MaterializedViewBuilderCore {
    as(qb) {
        if (typeof qb === 'function') {
            qb = qb(query_builders_1.queryBuilder);
        }
        const selectionProxy = new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        });
        const aliasedSelection = new Proxy(qb.getSelectedFields(), selectionProxy);
        return new Proxy(new PgMaterializedView({
            pgConfig: {
                with: this.config.with,
                using: this.config.using,
                tablespace: this.config.tablespace,
                withNoData: this.config.withNoData,
            },
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: aliasedSelection,
                query: qb.getSQL().inlineParams(),
            },
        }), selectionProxy);
    }
}
exports.MaterializedViewBuilder = MaterializedViewBuilder;
class ManualMaterializedViewBuilder extends MaterializedViewBuilderCore {
    constructor(name, columns, schema) {
        super(name, schema);
        this.columns = (0, utils_1.getTableColumns)((0, table_1.pgTable)(name, columns));
    }
    existing() {
        return new Proxy(new PgMaterializedView({
            pgConfig: undefined,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: undefined,
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
    as(query) {
        return new Proxy(new PgMaterializedView({
            pgConfig: undefined,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: query.inlineParams(),
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
}
exports.ManualMaterializedViewBuilder = ManualMaterializedViewBuilder;
class PgViewBase extends view_1.View {
}
exports.PgViewBase = PgViewBase;
exports.PgViewConfig = Symbol('PgViewConfig');
class PgView extends PgViewBase {
    constructor({ pgConfig, config }) {
        super(config);
        if (pgConfig) {
            this[exports.PgViewConfig] = {
                with: pgConfig.with,
            };
        }
    }
}
exports.PgView = PgView;
exports.PgMaterializedViewConfig = Symbol('PgMaterializedViewConfig');
class PgMaterializedView extends PgViewBase {
    constructor({ pgConfig, config }) {
        super(config);
        this[exports.PgMaterializedViewConfig] = {
            with: pgConfig === null || pgConfig === void 0 ? void 0 : pgConfig.with,
            using: pgConfig === null || pgConfig === void 0 ? void 0 : pgConfig.using,
            tablespace: pgConfig === null || pgConfig === void 0 ? void 0 : pgConfig.tablespace,
            withNoData: pgConfig === null || pgConfig === void 0 ? void 0 : pgConfig.withNoData,
        };
    }
}
exports.PgMaterializedView = PgMaterializedView;
/** @internal */
function pgViewWithSchema(name, selection, schema) {
    if (selection) {
        return new ManualViewBuilder(name, selection, schema);
    }
    return new ViewBuilder(name, schema);
}
exports.pgViewWithSchema = pgViewWithSchema;
/** @internal */
function pgMaterializedViewWithSchema(name, selection, schema) {
    if (selection) {
        return new ManualMaterializedViewBuilder(name, selection, schema);
    }
    return new MaterializedViewBuilder(name, schema);
}
exports.pgMaterializedViewWithSchema = pgMaterializedViewWithSchema;
function pgView(name, columns) {
    return pgViewWithSchema(name, columns, undefined);
}
exports.pgView = pgView;
function pgMaterializedView(name, columns) {
    return pgMaterializedViewWithSchema(name, columns, undefined);
}
exports.pgMaterializedView = pgMaterializedView;
