// drizzle-orm/postgres-js/session.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgresJsTransaction = exports.PostgresJsSession = exports.PostgresJsPreparedQuery = void 0;
const logger_1 = require("../logger");
const pg_core_1 = require("../pg-core");
const session_1 = require("../pg-core/session");
const sql_1 = require("../sql");
const utils_1 = require("../utils");
class PostgresJsPreparedQuery extends session_1.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name) {
        super();
        this.client = client;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.query = queryString;
    }
    execute(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.query, params);
        const { fields } = this;
        if (!fields) {
            return this.client.unsafe(this.query, params);
        }
        const result = this.client.unsafe(this.query, params).values();
        return result.then((result) => result.map((row) => (0, utils_1.mapResultRow)(fields, row, this.joinsNotNullableMap)));
    }
    all(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.query, params);
        return this.client.unsafe(this.query, params);
    }
    values(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.query, params);
        return this.client.unsafe(this.query, params).values();
    }
}
exports.PostgresJsPreparedQuery = PostgresJsPreparedQuery;
class PostgresJsSession extends session_1.PgSession {
    constructor(client, dialect, 
    /** @internal */
    options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.options = options;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new logger_1.NoopLogger();
    }
    prepareQuery(query, fields, name) {
        return new PostgresJsPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name);
    }
    query(query, params) {
        this.logger.logQuery(query, params);
        return this.client.unsafe(query, params).values();
    }
    queryObjects(query, params) {
        return this.client.unsafe(query, params);
    }
    transaction(transaction, config) {
        return this.client.begin((client) => __awaiter(this, void 0, void 0, function* () {
            const session = new PostgresJsSession(client, this.dialect, this.options);
            const tx = new PostgresJsTransaction(this.dialect, session);
            if (config) {
                yield tx.setTransaction(config);
            }
            return transaction(tx);
        }));
    }
}
exports.PostgresJsSession = PostgresJsSession;
class PostgresJsTransaction extends pg_core_1.PgTransaction {
    constructor(dialect, 
    /** @internal */
    session, nestedIndex = 0) {
        super(dialect, session, nestedIndex);
        this.session = session;
    }
    transaction(transaction) {
        return this.session.client.savepoint((client) => {
            const session = new PostgresJsSession(client, this.dialect, this.session.options);
            const tx = new PostgresJsTransaction(this.dialect, session);
            return transaction(tx);
        });
    }
}
exports.PostgresJsTransaction = PostgresJsTransaction;
