// drizzle-orm/sql/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fillPlaceholders = exports.placeholder = exports.Placeholder = exports.sql = exports.param = exports.Param = exports.noopMapper = exports.noopEncoder = exports.noopDecoder = exports.isDriverValueEncoder = exports.name = exports.Name = exports.SQL = exports.StringChunk = exports.isSQLWrapper = exports.FakePrimitiveParam = void 0;
const subquery_1 = require("../subquery");
const view_1 = require("../view");
const column_1 = require("../column");
const table_1 = require("../table");
/**
 * @internal
 * This class is used to indicate a primitive param value that is used in `sql` tag.
 * It is only used on type level and is never instantiated at runtime.
 * If you see a value of this type in the code, its runtime value is actually the primitive param value.
 */
class FakePrimitiveParam {
}
exports.FakePrimitiveParam = FakePrimitiveParam;
function isSQLWrapper(value) {
    return typeof value === 'object' && value !== null && 'getSQL' in value
        && typeof value.getSQL === 'function';
}
exports.isSQLWrapper = isSQLWrapper;
function mergeQueries(queries) {
    const result = { sql: '', params: [] };
    queries.forEach((query) => {
        var _a;
        result.sql += query.sql;
        result.params.push(...query.params);
        if (result.typings && ((_a = query.typings) === null || _a === void 0 ? void 0 : _a.length)) {
            result.typings.push(...query.typings);
        }
    });
    return result;
}
class StringChunk {
    constructor(value) {
        this.value = Array.isArray(value) ? value : [value];
    }
}
exports.StringChunk = StringChunk;
class SQL {
    constructor(queryChunks) {
        this.queryChunks = queryChunks;
        /** @internal */
        this.decoder = exports.noopDecoder;
        this.shouldInlineParams = false;
    }
    append(query) {
        this.queryChunks.push(...query.queryChunks);
        return this;
    }
    toQuery(config) {
        return this.buildQueryFromSourceParams(this.queryChunks, config);
    }
    buildQueryFromSourceParams(chunks, _config) {
        const config = Object.assign({}, _config, {
            inlineParams: _config.inlineParams || this.shouldInlineParams,
            paramStartIndex: _config.paramStartIndex || { value: 0 },
        });
        const { escapeName, escapeParam, escapeString, prepareTyping, inlineParams, paramStartIndex, } = config;
        return mergeQueries(chunks.map((chunk) => {
            if (chunk instanceof StringChunk) {
                return { sql: chunk.value.join(''), params: [] };
            }
            if (chunk instanceof Name) {
                return { sql: escapeName(chunk.value), params: [] };
            }
            if (typeof chunk === 'undefined') {
                return { sql: '', params: [] };
            }
            if (Array.isArray(chunk)) {
                const result = [new StringChunk('(')];
                chunk.forEach((p, i) => {
                    result.push(p);
                    if (i < chunk.length - 1) {
                        result.push(new StringChunk(', '));
                    }
                });
                result.push(new StringChunk(')'));
                return this.buildQueryFromSourceParams(result, config);
            }
            if (chunk instanceof SQL) {
                return this.buildQueryFromSourceParams(chunk.queryChunks, Object.assign(Object.assign({}, config), { inlineParams: inlineParams || chunk.shouldInlineParams }));
            }
            if (chunk instanceof table_1.Table) {
                const schemaName = chunk[table_1.Table.Symbol.Schema];
                const tableName = chunk[table_1.Table.Symbol.Name];
                return {
                    sql: typeof schemaName !== 'undefined'
                        ? escapeName(schemaName) + '.' + escapeName(tableName)
                        : escapeName(tableName),
                    params: [],
                };
            }
            if (chunk instanceof column_1.Column) {
                return { sql: escapeName(chunk.table[table_1.Table.Symbol.Name]) + '.' + escapeName(chunk.name), params: [] };
            }
            if (chunk instanceof view_1.View) {
                const schemaName = chunk[view_1.ViewBaseConfig].schema;
                const viewName = chunk[view_1.ViewBaseConfig].name;
                return {
                    sql: typeof schemaName !== 'undefined'
                        ? escapeName(schemaName) + '.' + escapeName(viewName)
                        : escapeName(viewName),
                    params: [],
                };
            }
            if (chunk instanceof Param) {
                const mappedValue = chunk.value === null ? null : chunk.encoder.mapToDriverValue(chunk.value);
                if (mappedValue instanceof SQL) {
                    return this.buildQueryFromSourceParams([mappedValue], config);
                }
                if (inlineParams) {
                    return { sql: this.mapInlineParam(mappedValue, config), params: [] };
                }
                let typings;
                if (typeof prepareTyping !== 'undefined') {
                    typings = [prepareTyping(chunk.encoder)];
                }
                return { sql: escapeParam(paramStartIndex.value++, mappedValue), params: [mappedValue], typings };
            }
            if (chunk instanceof SQL.Aliased && typeof chunk.fieldAlias !== 'undefined') {
                return { sql: escapeName(chunk.fieldAlias), params: [] };
            }
            if (chunk instanceof subquery_1.Subquery) {
                if (chunk[subquery_1.SubqueryConfig].isWith) {
                    return { sql: escapeName(chunk[subquery_1.SubqueryConfig].alias), params: [] };
                }
                return this.buildQueryFromSourceParams([
                    new StringChunk('('),
                    chunk[subquery_1.SubqueryConfig].sql,
                    new StringChunk(') '),
                    new Name(chunk[subquery_1.SubqueryConfig].alias),
                ], config);
            }
            if (isSQLWrapper(chunk)) {
                return this.buildQueryFromSourceParams([
                    new StringChunk('('),
                    chunk.getSQL(),
                    new StringChunk(')'),
                ], config);
            }
            if (inlineParams) {
                return { sql: this.mapInlineParam(chunk, config), params: [] };
            }
            return { sql: escapeParam(paramStartIndex.value++, chunk), params: [chunk] };
        }));
    }
    mapInlineParam(chunk, { escapeString }) {
        if (chunk === null) {
            return 'null';
        }
        if (typeof chunk === 'number' || typeof chunk === 'boolean') {
            return chunk.toString();
        }
        if (typeof chunk === 'string') {
            return escapeString(chunk);
        }
        if (typeof chunk === 'object') {
            const mappedValueAsString = chunk.toString();
            if (mappedValueAsString === '[object Object]') {
                return escapeString(JSON.stringify(chunk));
            }
            return escapeString(mappedValueAsString);
        }
        throw new Error('Unexpected param value: ' + chunk);
    }
    getSQL() {
        return this;
    }
    as(alias) {
        // TODO: remove with deprecated overloads
        if (typeof alias === 'undefined') {
            return this;
        }
        return new SQL.Aliased(this, alias);
    }
    mapWith(decoder) {
        if (typeof decoder === 'function') {
            this.decoder = { mapFromDriverValue: decoder };
        }
        else {
            this.decoder = decoder;
        }
        return this;
    }
    inlineParams() {
        this.shouldInlineParams = true;
        return this;
    }
}
exports.SQL = SQL;
/**
 * Any DB name (table, column, index etc.)
 */
class Name {
    constructor(value) {
        this.value = value;
    }
}
exports.Name = Name;
/**
 * Any DB name (table, column, index etc.)
 */
function name(value) {
    return new Name(value);
}
exports.name = name;
function isDriverValueEncoder(value) {
    return typeof value === 'object' && value !== null && 'mapToDriverValue' in value
        && typeof value.mapToDriverValue === 'function';
}
exports.isDriverValueEncoder = isDriverValueEncoder;
exports.noopDecoder = {
    mapFromDriverValue: (value) => value,
};
exports.noopEncoder = {
    mapToDriverValue: (value) => value,
};
exports.noopMapper = Object.assign(Object.assign({}, exports.noopDecoder), exports.noopEncoder);
/** Parameter value that is optionally bound to an encoder (for example, a column). */
class Param {
    /**
     * @param value - Parameter value
     * @param encoder - Encoder to convert the value to a driver parameter
     */
    constructor(value, encoder = exports.noopEncoder) {
        this.value = value;
        this.encoder = encoder;
    }
}
exports.Param = Param;
function param(value, encoder) {
    return new Param(value, encoder);
}
exports.param = param;
/*
    The type of `params` is specified as `SQLSourceParam[]`, but that's slightly incorrect -
    in runtime, users won't pass `FakePrimitiveParam` instances as `params` - they will pass primitive values
    which will be wrapped in `Param` using `buildChunksFromParam(...)`. That's why the overload
    specify `params` as `any[]` and not as `SQLSourceParam[]`. This type is used to make our lives easier and
    the type checker happy.
*/
function sql(strings, ...params) {
    const queryChunks = [];
    if (params.length > 0 || (strings.length > 0 && strings[0] !== '')) {
        queryChunks.push(new StringChunk(strings[0]));
    }
    params.forEach((param, paramIndex) => {
        queryChunks.push(param);
        queryChunks.push(new StringChunk(strings[paramIndex + 1]));
    });
    return new SQL(queryChunks);
}
exports.sql = sql;
(function (sql) {
    function empty() {
        return new SQL([]);
    }
    sql.empty = empty;
    function fromList(list) {
        return new SQL(list);
    }
    sql.fromList = fromList;
    /**
     * Convenience function to create an SQL query from a raw string.
     * @param str The raw SQL query string.
     */
    function raw(str) {
        return new SQL([new StringChunk(str)]);
    }
    sql.raw = raw;
    /**
     * Convenience function to join a list of SQL chunks with a separator.
     */
    function join(chunks, separator) {
        const result = [];
        for (let i = 0; i < chunks.length; i++) {
            if (i > 0) {
                result.push(separator);
            }
            result.push(chunks[i]);
        }
        return sql.fromList(result);
    }
    sql.join = join;
})(sql = exports.sql || (exports.sql = {}));
(function (SQL) {
    class Aliased {
        constructor(sql, fieldAlias) {
            this.sql = sql;
            this.fieldAlias = fieldAlias;
            /** @internal */
            this.isSelectionField = false;
        }
        getSQL() {
            return this.sql;
        }
        /** @internal */
        clone() {
            return new Aliased(this.sql, this.fieldAlias);
        }
    }
    SQL.Aliased = Aliased;
})(SQL = exports.SQL || (exports.SQL = {}));
class Placeholder {
    constructor(name) {
        this.name = name;
    }
}
exports.Placeholder = Placeholder;
function placeholder(name) {
    return new Placeholder(name);
}
exports.placeholder = placeholder;
function fillPlaceholders(params, values) {
    return params.map((p) => {
        if (p instanceof Placeholder) {
            if (!(p.name in values)) {
                throw new Error(`No value for placeholder "${p.name}" was provided`);
            }
            return values[p.name];
        }
        return p;
    });
}
exports.fillPlaceholders = fillPlaceholders;
