// drizzle-orm/d1/session.js
"use strict";
/// <reference types="@cloudflare/workers-types" />
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreparedQuery = exports.D1Transaction = exports.SQLiteD1Session = void 0;
const logger_1 = require("../logger");
const sql_1 = require("../sql");
const sql_2 = require("../sql");
const sqlite_core_1 = require("../sqlite-core");
const session_1 = require("../sqlite-core/session");
const utils_1 = require("../utils");
class SQLiteD1Session extends session_1.SQLiteSession {
    constructor(client, dialect, options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new logger_1.NoopLogger();
    }
    prepareQuery(query, fields) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields);
    }
    transaction(transaction, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const tx = new D1Transaction(this.dialect, this);
            yield this.run(sql_1.sql.raw(`begin${(config === null || config === void 0 ? void 0 : config.behavior) ? ' ' + config.behavior : ''}`));
            try {
                const result = yield transaction(tx);
                yield this.run((0, sql_1.sql) `commit`);
                return result;
            }
            catch (err) {
                yield this.run((0, sql_1.sql) `rollback`);
                throw err;
            }
        });
    }
}
exports.SQLiteD1Session = SQLiteD1Session;
class D1Transaction extends sqlite_core_1.SQLiteTransaction {
    transaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            const savepointName = `sp${this.nestedIndex}`;
            const tx = new D1Transaction(this.dialect, this.session, this.nestedIndex + 1);
            yield this.session.run(sql_1.sql.raw(`savepoint ${savepointName}`));
            try {
                const result = yield transaction(tx);
                yield this.session.run(sql_1.sql.raw(`release savepoint ${savepointName}`));
                return result;
            }
            catch (err) {
                yield this.session.run(sql_1.sql.raw(`rollback to savepoint ${savepointName}`));
                throw err;
            }
        });
    }
}
exports.D1Transaction = D1Transaction;
class PreparedQuery extends session_1.PreparedQuery {
    constructor(stmt, queryString, params, logger, fields) {
        super();
        this.stmt = stmt;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
    }
    run(placeholderValues) {
        const params = (0, sql_2.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.bind(...params).run();
    }
    all(placeholderValues) {
        const { fields } = this;
        if (fields) {
            return this.values(placeholderValues).then((values) => values.map((row) => (0, utils_1.mapResultRow)(fields, row, this.joinsNotNullableMap)));
        }
        const params = (0, sql_2.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.bind(...params).all().then(({ results }) => results);
    }
    get(placeholderValues) {
        // TODO: implement using stmt.get()
        return this.all(placeholderValues).then((rows) => rows[0]);
    }
    values(placeholderValues) {
        const params = (0, sql_2.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.bind(...params).raw();
    }
}
exports.PreparedQuery = PreparedQuery;
