// drizzle-orm/pg-core/columns/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgArray = exports.PgArrayBuilder = void 0;
const utils_1 = require("../utils");
const common_1 = require("./common");
class PgArrayBuilder extends common_1.PgColumnBuilder {
    constructor(name, baseBuilder, size) {
        super(name);
        this.config.baseBuilder = baseBuilder;
        this.config.size = size;
    }
    /** @internal */
    build(table) {
        const baseColumn = this.config.baseBuilder.build(table);
        return new PgArray(table, this.config, baseColumn);
    }
}
exports.PgArrayBuilder = PgArrayBuilder;
class PgArray extends common_1.PgColumn {
    constructor(table, config, baseColumn, range) {
        super(table, config);
        this.baseColumn = baseColumn;
        this.range = range;
        this.size = config.size;
    }
    getSQLType() {
        return `${this.baseColumn.getSQLType()}[${typeof this.size === 'number' ? this.size : ''}]`;
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            // Thank you node-postgres for not parsing enum arrays
            value = (0, utils_1.parsePgArray)(value);
        }
        return value.map((v) => this.baseColumn.mapFromDriverValue(v));
    }
    mapToDriverValue(value) {
        const a = value.map((v) => v === null ? null : this.baseColumn.mapToDriverValue(v));
        return (0, utils_1.makePgArray)(a);
    }
}
exports.PgArray = PgArray;
