// drizzle-orm/pg-core/columns/char.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.char = exports.PgChar = exports.PgCharBuilder = void 0;
const common_1 = require("./common");
class PgCharBuilder extends common_1.PgColumnBuilder {
    constructor(name, config) {
        var _a;
        super(name);
        this.config.length = config.length;
        this.config.enumValues = ((_a = config.enum) !== null && _a !== void 0 ? _a : []);
    }
    /** @internal */
    build(table) {
        return new PgChar(table, this.config);
    }
}
exports.PgCharBuilder = PgCharBuilder;
class PgChar extends common_1.PgColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = this.config.enumValues;
    }
    getSQLType() {
        return typeof this.length !== 'undefined' ? `char(${this.length})` : `char`;
    }
}
exports.PgChar = PgChar;
function char(name, config = {}) {
    return new PgCharBuilder(name, config);
}
exports.char = char;
