// drizzle-orm/pg-core/columns/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customType = exports.PgCustomColumn = exports.PgCustomColumnBuilder = void 0;
const common_1 = require("./common");
class PgCustomColumnBuilder extends common_1.PgColumnBuilder {
    constructor(name, fieldConfig, customTypeParams) {
        super(name);
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new PgCustomColumn(table, this.config);
    }
}
exports.PgCustomColumnBuilder = PgCustomColumnBuilder;
class PgCustomColumn extends common_1.PgColumn {
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        if (typeof this.mapFrom === 'function') {
            return this.mapFrom(value);
        }
        else {
            return value;
        }
    }
    mapToDriverValue(value) {
        if (typeof this.mapTo === 'function') {
            return this.mapTo(value);
        }
        else {
            return value;
        }
    }
}
exports.PgCustomColumn = PgCustomColumn;
/**
 * Custom pg database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new PgCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}
exports.customType = customType;
