// drizzle-orm/pg-core/columns/varchar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.varchar = exports.PgVarchar = exports.PgVarcharBuilder = void 0;
const common_1 = require("./common");
class PgVarcharBuilder extends common_1.PgColumnBuilder {
    constructor(name, config) {
        var _a;
        super(name);
        this.config.length = config.length;
        this.config.enumValues = ((_a = config.enum) !== null && _a !== void 0 ? _a : []);
    }
    /** @internal */
    build(table) {
        return new PgVarchar(table, this.config);
    }
}
exports.PgVarcharBuilder = PgVarcharBuilder;
class PgVarchar extends common_1.PgColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = this.config.enumValues;
    }
    getSQLType() {
        return typeof this.length !== 'undefined' ? `varchar(${this.length})` : `varchar`;
    }
}
exports.PgVarchar = PgVarchar;
function varchar(name, config = {}) {
    return new PgVarcharBuilder(name, config);
}
exports.varchar = varchar;
