// drizzle-orm/libsql/session.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreparedQuery = exports.LibSQLTransaction = exports.LibSQLSession = void 0;
const logger_1 = require("../logger");
const sql_1 = require("../sql");
const sqlite_core_1 = require("../sqlite-core");
const session_1 = require("../sqlite-core/session");
const utils_1 = require("../utils");
class LibSQLSession extends session_1.SQLiteSession {
    constructor(client, dialect, options, tx) {
        var _a;
        super(dialect);
        this.client = client;
        this.options = options;
        this.tx = tx;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new logger_1.NoopLogger();
    }
    prepareQuery(query, fields) {
        return new PreparedQuery(this.client, query.sql, query.params, this.logger, fields, this.tx);
    }
    /*override */ batch(queries) {
        const builtQueries = queries.map((query) => {
            const builtQuery = this.dialect.sqlToQuery(query);
            return { sql: builtQuery.sql, args: builtQuery.params };
        });
        return this.client.batch(builtQueries);
    }
    transaction(transaction, config) {
        return __awaiter(this, void 0, void 0, function* () {
            // TODO: support transaction behavior
            const libsqlTx = yield this.client.transaction();
            const session = new LibSQLSession(this.client, this.dialect, this.options, libsqlTx);
            const tx = new LibSQLTransaction(this.dialect, session);
            try {
                const result = yield transaction(tx);
                yield libsqlTx.commit();
                return result;
            }
            catch (err) {
                yield libsqlTx.rollback();
                throw err;
            }
        });
    }
}
exports.LibSQLSession = LibSQLSession;
class LibSQLTransaction extends sqlite_core_1.SQLiteTransaction {
    transaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            const savepointName = `sp${this.nestedIndex}`;
            const tx = new LibSQLTransaction(this.dialect, this.session, this.nestedIndex + 1);
            yield this.session.run(sql_1.sql.raw(`savepoint ${savepointName}`));
            try {
                const result = yield transaction(tx);
                yield this.session.run(sql_1.sql.raw(`release savepoint ${savepointName}`));
                return result;
            }
            catch (err) {
                yield this.session.run(sql_1.sql.raw(`rollback to savepoint ${savepointName}`));
                throw err;
            }
        });
    }
}
exports.LibSQLTransaction = LibSQLTransaction;
class PreparedQuery extends session_1.PreparedQuery {
    constructor(client, queryString, params, logger, fields, tx) {
        super();
        this.client = client;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.tx = tx;
    }
    run(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const stmt = { sql: this.queryString, args: params };
        return this.tx ? this.tx.execute(stmt) : this.client.execute(stmt);
    }
    all(placeholderValues) {
        const { fields } = this;
        if (fields) {
            const values = this.values(placeholderValues);
            return values.then((rows) => rows.map((row) => {
                return (0, utils_1.mapResultRow)(fields, Array.prototype.slice.call(row).map(normalizeFieldValue), this.joinsNotNullableMap);
            }));
        }
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const stmt = { sql: this.queryString, args: params };
        return (this.tx ? this.tx.execute(stmt) : this.client.execute(stmt)).then(({ rows }) => rows.map(normalizeRow));
    }
    get(placeholderValues) {
        return this.all(placeholderValues).then((rows) => rows[0]);
    }
    values(placeholderValues) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues !== null && placeholderValues !== void 0 ? placeholderValues : {});
        this.logger.logQuery(this.queryString, params);
        const stmt = { sql: this.queryString, args: params };
        return (this.tx ? this.tx.execute(stmt) : this.client.execute(stmt)).then(({ rows }) => rows);
    }
}
exports.PreparedQuery = PreparedQuery;
function normalizeRow(obj) {
    // The libSQL node-sqlite3 compatibility wrapper returns rows
    // that can be accessed both as objects and arrays. Let's
    // turn them into objects what's what other SQLite drivers
    // do.
    return Object.keys(obj).reduce((acc, key) => {
        if (obj.propertyIsEnumerable(key)) {
            acc[key] = obj[key];
        }
        return acc;
    }, {});
}
function normalizeFieldValue(value) {
    if (value instanceof ArrayBuffer) {
        if (typeof Buffer !== 'undefined') {
            if (!(value instanceof Buffer)) {
                return Buffer.from(value);
            }
            return value;
        }
        if (typeof TextDecoder !== 'undefined') {
            return new TextDecoder().decode(value);
        }
        throw new Error('TextDecoder is not available. Please provide either Buffer or TextDecoder polyfill.');
    }
    return value;
}
