// drizzle-orm/mysql-core/query-builders/select.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySqlSelect = exports.MySqlSelectQueryBuilder = exports.MySqlSelectBuilder = void 0;
const view_1 = require("../view");
const query_builder_1 = require("../../query-builders/query-builder");
const query_promise_1 = require("../../query-promise");
const sql_1 = require("../../sql");
const subquery_1 = require("../../subquery");
const table_1 = require("../../table");
const utils_1 = require("../../utils");
const utils_2 = require("../../utils");
const view_2 = require("../../view");
class MySqlSelectBuilder {
    constructor(fields, session, dialect, withList = []) {
        this.fields = fields;
        this.session = session;
        this.dialect = dialect;
        this.withList = withList;
    }
    from(source) {
        const isPartialSelect = !!this.fields;
        let fields;
        if (this.fields) {
            fields = this.fields;
        }
        else if (source instanceof subquery_1.Subquery) {
            // This is required to use the proxy handler to get the correct field values from the subquery
            fields = Object.fromEntries(Object.keys(source[subquery_1.SubqueryConfig].selection).map((key) => [key, source[key]]));
        }
        else if (source instanceof view_1.MySqlViewBase) {
            fields = source[view_2.ViewBaseConfig].selectedFields;
        }
        else if (source instanceof sql_1.SQL) {
            fields = {};
        }
        else {
            fields = (0, utils_1.getTableColumns)(source);
        }
        return new MySqlSelect(source, fields, isPartialSelect, this.session, this.dialect, this.withList);
    }
}
exports.MySqlSelectBuilder = MySqlSelectBuilder;
class MySqlSelectQueryBuilder extends query_builder_1.QueryBuilder {
    constructor(table, fields, isPartialSelect, session, dialect, withList) {
        super();
        this.isPartialSelect = isPartialSelect;
        this.session = session;
        this.dialect = dialect;
        this.leftJoin = this.createJoin('left');
        this.rightJoin = this.createJoin('right');
        this.innerJoin = this.createJoin('inner');
        this.fullJoin = this.createJoin('full');
        this.config = {
            withList,
            table,
            fields: Object.assign({}, fields),
            joins: [],
            orderBy: [],
            groupBy: [],
        };
        this._ = {
            selectedFields: fields,
        };
        this.tableName = table instanceof subquery_1.Subquery
            ? table[subquery_1.SubqueryConfig].alias
            : table instanceof view_1.MySqlViewBase
                ? table[view_2.ViewBaseConfig].name
                : table instanceof sql_1.SQL
                    ? undefined
                    : table[table_1.Table.Symbol.BaseName];
        this.joinsNotNullableMap = typeof this.tableName === 'string' ? { [this.tableName]: true } : {};
    }
    createJoin(joinType) {
        return (table, on) => {
            const baseTableName = this.tableName;
            const tableName = table instanceof subquery_1.Subquery
                ? table[subquery_1.SubqueryConfig].alias
                : table instanceof view_2.View
                    ? table[view_2.ViewBaseConfig].name
                    : table instanceof sql_1.SQL
                        ? undefined
                        : table[table_1.Table.Symbol.Name];
            if (typeof tableName === 'string' && this.config.joins.some((join) => join.alias === tableName)) {
                throw new Error(`Alias "${tableName}" is already used in this query`);
            }
            if (!this.isPartialSelect) {
                // If this is the first join and this is not a partial select and we're not selecting from raw SQL, "move" the fields from the main table to the nested object
                if (Object.keys(this.joinsNotNullableMap).length === 1 && typeof baseTableName === 'string') {
                    this.config.fields = {
                        [baseTableName]: this.config.fields,
                    };
                }
                if (typeof tableName === 'string' && !(table instanceof sql_1.SQL)) {
                    const selection = table instanceof subquery_1.Subquery
                        ? table[subquery_1.SubqueryConfig].selection
                        : table instanceof view_2.View
                            ? table[view_2.ViewBaseConfig].selectedFields
                            : table[table_1.Table.Symbol.Columns];
                    this.config.fields[tableName] = selection;
                }
            }
            if (typeof on === 'function') {
                on = on(new Proxy(this.config.fields, new subquery_1.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
            }
            this.config.joins.push({ on, table, joinType, alias: tableName });
            if (typeof tableName === 'string') {
                switch (joinType) {
                    case 'left':
                        this.joinsNotNullableMap[tableName] = false;
                        break;
                    case 'right':
                        this.joinsNotNullableMap = Object.fromEntries(Object.entries(this.joinsNotNullableMap).map(([key]) => [key, false]));
                        this.joinsNotNullableMap[tableName] = true;
                        break;
                    case 'inner':
                        this.joinsNotNullableMap[tableName] = true;
                        break;
                    case 'full':
                        this.joinsNotNullableMap = Object.fromEntries(Object.entries(this.joinsNotNullableMap).map(([key]) => [key, false]));
                        this.joinsNotNullableMap[tableName] = false;
                        break;
                }
            }
            return this;
        };
    }
    where(where) {
        if (typeof where === 'function') {
            where = where(new Proxy(this.config.fields, new subquery_1.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
        }
        this.config.where = where;
        return this;
    }
    having(having) {
        if (typeof having === 'function') {
            having = having(new Proxy(this.config.fields, new subquery_1.SelectionProxyHandler({ sqlAliasedBehavior: 'sql', sqlBehavior: 'sql' })));
        }
        this.config.having = having;
        return this;
    }
    groupBy(...columns) {
        if (typeof columns[0] === 'function') {
            const groupBy = columns[0](new Proxy(this.config.fields, new subquery_1.SelectionProxyHandler({ sqlAliasedBehavior: 'alias', sqlBehavior: 'sql' })));
            this.config.groupBy = Array.isArray(groupBy) ? groupBy : [groupBy];
        }
        else {
            this.config.groupBy = columns;
        }
        return this;
    }
    orderBy(...columns) {
        if (typeof columns[0] === 'function') {
            const orderBy = columns[0](new Proxy(this.config.fields, new subquery_1.SelectionProxyHandler({ sqlAliasedBehavior: 'alias', sqlBehavior: 'sql' })));
            this.config.orderBy = Array.isArray(orderBy) ? orderBy : [orderBy];
        }
        else {
            this.config.orderBy = columns;
        }
        return this;
    }
    limit(limit) {
        this.config.limit = limit;
        return this;
    }
    offset(offset) {
        this.config.offset = offset;
        return this;
    }
    for(strength, config = {}) {
        this.config.lockingClause = { strength, config };
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildSelectQuery(this.config);
    }
    toSQL() {
        const _a = this.dialect.sqlToQuery(this.getSQL()), { typings } = _a, rest = __rest(_a, ["typings"]);
        return rest;
    }
    as(alias) {
        return new Proxy(new subquery_1.Subquery(this.getSQL(), this.config.fields, alias), new subquery_1.SelectionProxyHandler({ alias, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
    }
}
exports.MySqlSelectQueryBuilder = MySqlSelectQueryBuilder;
class MySqlSelect extends MySqlSelectQueryBuilder {
    constructor() {
        super(...arguments);
        this.execute = (placeholderValues) => {
            return this._prepare().execute(placeholderValues);
        };
    }
    _prepare(name) {
        if (!this.session) {
            throw new Error('Cannot execute a query on a query builder. Please use a database instance instead.');
        }
        const fieldsList = (0, utils_2.orderSelectedFields)(this.config.fields);
        const query = this.session.prepareQuery(this.dialect.sqlToQuery(this.getSQL()), fieldsList, name);
        query.joinsNotNullableMap = this.joinsNotNullableMap;
        return query;
    }
    prepare(name) {
        return this._prepare(name);
    }
}
exports.MySqlSelect = MySqlSelect;
(0, utils_1.applyMixins)(MySqlSelect, [query_promise_1.QueryPromise]);
