// drizzle-orm/planetscale-serverless/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.PlanetscaleDriver = void 0;
const logger_1 = require("../logger");
const db_1 = require("../mysql-core/db");
const dialect_1 = require("../mysql-core/dialect");
const session_1 = require("./session");
class PlanetscaleDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
    }
    createSession() {
        return new session_1.PlanetscaleSession(this.client, this.dialect, undefined, { logger: this.options.logger });
    }
}
exports.PlanetscaleDriver = PlanetscaleDriver;
function drizzle(client, config = {}) {
    const dialect = new dialect_1.MySqlDialect();
    let logger;
    if (config.logger === true) {
        logger = new logger_1.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    const driver = new PlanetscaleDriver(client, dialect, { logger });
    const session = driver.createSession();
    return new db_1.MySqlDatabase(dialect, session);
}
exports.drizzle = drizzle;
