// drizzle-orm/table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTableName = exports.Table = exports.BaseName = exports.OriginalName = exports.Columns = exports.Schema = exports.TableName = void 0;
/** @internal */
exports.TableName = Symbol('Name');
/** @internal */
exports.Schema = Symbol('Schema');
/** @internal */
exports.Columns = Symbol('Columns');
/** @internal */
exports.OriginalName = Symbol('OriginalName');
/** @internal */
exports.BaseName = Symbol('BaseName');
class Table {
    constructor(name, schema, baseName) {
        this[exports.TableName] = this[exports.OriginalName] = name;
        this[exports.Schema] = schema;
        this[exports.BaseName] = baseName;
    }
}
/** @internal */
Table.Symbol = {
    Name: exports.TableName,
    Schema: exports.Schema,
    OriginalName: exports.OriginalName,
    Columns: exports.Columns,
    BaseName: exports.BaseName,
};
exports.Table = Table;
function getTableName(table) {
    return table[exports.TableName];
}
exports.getTableName = getTableName;
