// drizzle-orm/mysql-core/columns/char.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.char = exports.MySqlChar = exports.MySqlCharBuilder = void 0;
const common_1 = require("./common");
class MySqlCharBuilder extends common_1.MySqlColumnBuilder {
    constructor(name, config) {
        super(name);
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlChar(table, this.config);
    }
}
exports.MySqlCharBuilder = MySqlCharBuilder;
class MySqlChar extends common_1.MySqlColumn {
    constructor() {
        var _a;
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = ((_a = this.config.enum) !== null && _a !== void 0 ? _a : []);
    }
    getSQLType() {
        return typeof this.length !== 'undefined' ? `char(${this.length})` : `char`;
    }
}
exports.MySqlChar = MySqlChar;
function char(name, config = {}) {
    return new MySqlCharBuilder(name, config);
}
exports.char = char;
