// drizzle-orm/mysql-core/columns/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customType = exports.MySqlCustomColumn = exports.MySqlCustomColumnBuilder = void 0;
const common_1 = require("./common");
class MySqlCustomColumnBuilder extends common_1.MySqlColumnBuilder {
    constructor(name, fieldConfig, customTypeParams) {
        super(name);
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new MySqlCustomColumn(table, this.config);
    }
}
exports.MySqlCustomColumnBuilder = MySqlCustomColumnBuilder;
class MySqlCustomColumn extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        if (typeof this.mapFrom === 'function') {
            return this.mapFrom(value);
        }
        else {
            return value;
        }
    }
    mapToDriverValue(value) {
        if (typeof this.mapTo === 'function') {
            return this.mapTo(value);
        }
        else {
            return value;
        }
    }
}
exports.MySqlCustomColumn = MySqlCustomColumn;
/**
 * Custom mysql database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new MySqlCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}
exports.customType = customType;
