// drizzle-orm/mysql-core/columns/date.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.date = exports.MySqlDateString = exports.MySqlDateStringBuilder = exports.MySqlDate = exports.MySqlDateBuilder = void 0;
const common_1 = require("./common");
class MySqlDateBuilder extends common_1.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlDate(table, this.config);
    }
}
exports.MySqlDateBuilder = MySqlDateBuilder;
class MySqlDate extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
    mapFromDriverValue(value) {
        return new Date(value);
    }
}
exports.MySqlDate = MySqlDate;
class MySqlDateStringBuilder extends common_1.MySqlColumnBuilder {
    /** @internal */
    build(table) {
        return new MySqlDateString(table, this.config);
    }
}
exports.MySqlDateStringBuilder = MySqlDateStringBuilder;
class MySqlDateString extends common_1.MySqlColumn {
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
}
exports.MySqlDateString = MySqlDateString;
function date(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlDateStringBuilder(name);
    }
    return new MySqlDateBuilder(name);
}
exports.date = date;
