// drizzle-orm/mysql-core/columns/real.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.real = exports.MySqlReal = exports.MySqlRealBuilder = void 0;
const common_1 = require("./common");
class MySqlRealBuilder extends common_1.MySqlColumnBuilderWithAutoIncrement {
    constructor(name, config) {
        super(name);
        this.config.precision = config === null || config === void 0 ? void 0 : config.precision;
        this.config.scale = config === null || config === void 0 ? void 0 : config.scale;
    }
    /** @internal */
    build(table) {
        return new MySqlReal(table, this.config);
    }
}
exports.MySqlRealBuilder = MySqlRealBuilder;
class MySqlReal extends common_1.MySqlColumnWithAutoIncrement {
    constructor() {
        super(...arguments);
        this.precision = this.config.precision;
        this.scale = this.config.scale;
    }
    getSQLType() {
        if (typeof this.precision !== 'undefined' && typeof this.scale !== 'undefined') {
            return `real(${this.precision}, ${this.scale})`;
        }
        else if (typeof this.precision === 'undefined') {
            return 'real';
        }
        else {
            return `real(${this.precision})`;
        }
    }
}
exports.MySqlReal = MySqlReal;
function real(name, config = {}) {
    return new MySqlRealBuilder(name, config);
}
exports.real = real;
