// drizzle-orm/mysql-core/columns/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.longtext = exports.mediumtext = exports.tinytext = exports.text = exports.MySqlText = exports.MySqlTextBuilder = void 0;
const common_1 = require("./common");
class MySqlTextBuilder extends common_1.MySqlColumnBuilder {
    constructor(name, textType, config) {
        super(name);
        this.config.textType = textType;
        this.config.enumValues = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlText(table, this.config);
    }
}
exports.MySqlTextBuilder = MySqlTextBuilder;
class MySqlText extends common_1.MySqlColumn {
    constructor() {
        var _a;
        super(...arguments);
        this.textType = this.config.textType;
        this.enumValues = ((_a = this.config.enumValues) !== null && _a !== void 0 ? _a : []);
    }
    getSQLType() {
        return this.textType;
    }
}
exports.MySqlText = MySqlText;
function text(name, config = {}) {
    return new MySqlTextBuilder(name, 'text', config);
}
exports.text = text;
function tinytext(name, config = {}) {
    return new MySqlTextBuilder(name, 'tinytext', config);
}
exports.tinytext = tinytext;
function mediumtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'mediumtext', config);
}
exports.mediumtext = mediumtext;
function longtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'longtext', config);
}
exports.longtext = longtext;
