// drizzle-orm/mysql-core/columns/timestamp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timestamp = exports.MySqlTimestampString = exports.MySqlTimestampStringBuilder = exports.MySqlTimestamp = exports.MySqlTimestampBuilder = void 0;
const date_common_1 = require("./date.common");
class MySqlTimestampBuilder extends date_common_1.MySqlDateColumnBaseBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config === null || config === void 0 ? void 0 : config.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestamp(table, this.config);
    }
}
exports.MySqlTimestampBuilder = MySqlTimestampBuilder;
class MySqlTimestamp extends date_common_1.MySqlDateBaseColumn {
    constructor() {
        super(...arguments);
        this.fsp = this.config.fsp;
    }
    getSQLType() {
        const precision = typeof this.fsp !== 'undefined' ? `(${this.fsp})` : '';
        return `timestamp${precision}`;
    }
    mapFromDriverValue(value) {
        return new Date(value + '+0000');
    }
    mapToDriverValue(value) {
        return value.toISOString().slice(0, 19).replace('T', ' ');
    }
}
exports.MySqlTimestamp = MySqlTimestamp;
class MySqlTimestampStringBuilder extends date_common_1.MySqlDateColumnBaseBuilder {
    constructor(name, config) {
        super(name);
        this.config.fsp = config === null || config === void 0 ? void 0 : config.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestampString(table, this.config);
    }
}
exports.MySqlTimestampStringBuilder = MySqlTimestampStringBuilder;
class MySqlTimestampString extends date_common_1.MySqlDateBaseColumn {
    constructor() {
        super(...arguments);
        this.fsp = this.config.fsp;
    }
    getSQLType() {
        const precision = typeof this.fsp !== 'undefined' ? `(${this.fsp})` : '';
        return `timestamp${precision}`;
    }
}
exports.MySqlTimestampString = MySqlTimestampString;
function timestamp(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlTimestampStringBuilder(name, config);
    }
    return new MySqlTimestampBuilder(name, config);
}
exports.timestamp = timestamp;
