// drizzle-orm/mysql-core/columns/varbinary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.varbinary = exports.MySqlVarBinary = exports.MySqlVarBinaryBuilder = void 0;
const common_1 = require("./common");
class MySqlVarBinaryBuilder extends common_1.MySqlColumnBuilder {
    /** @internal */
    constructor(name, config) {
        super(name);
        this.config.length = config === null || config === void 0 ? void 0 : config.length;
    }
    /** @internal */
    build(table) {
        return new MySqlVarBinary(table, this.config);
    }
}
exports.MySqlVarBinaryBuilder = MySqlVarBinaryBuilder;
class MySqlVarBinary extends common_1.MySqlColumn {
    constructor() {
        super(...arguments);
        this.length = this.config.length;
    }
    getSQLType() {
        return typeof this.length !== 'undefined' ? `varbinary(${this.length})` : `varbinary`;
    }
}
exports.MySqlVarBinary = MySqlVarBinary;
function varbinary(name, options) {
    return new MySqlVarBinaryBuilder(name, options);
}
exports.varbinary = varbinary;
