// drizzle-orm/mysql-core/columns/varchar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.varchar = exports.MySqlVarChar = exports.MySqlVarCharBuilder = void 0;
const common_1 = require("./common");
class MySqlVarCharBuilder extends common_1.MySqlColumnBuilder {
    /** @internal */
    constructor(name, config) {
        super(name);
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlVarChar(table, this.config);
    }
}
exports.MySqlVarCharBuilder = MySqlVarCharBuilder;
class MySqlVarChar extends common_1.MySqlColumn {
    constructor() {
        var _a;
        super(...arguments);
        this.length = this.config.length;
        this.enumValues = ((_a = this.config.enum) !== null && _a !== void 0 ? _a : []);
    }
    getSQLType() {
        return typeof this.length !== 'undefined' ? `varchar(${this.length})` : `varchar`;
    }
}
exports.MySqlVarChar = MySqlVarChar;
function varchar(name, config) {
    return new MySqlVarCharBuilder(name, config);
}
exports.varchar = varchar;
