// drizzle-orm/sqlite-core/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteTransaction = exports.BaseSQLiteDatabase = void 0;
const errors_1 = require("../errors");
const query_builders_1 = require("./query-builders");
const subquery_1 = require("../subquery");
class BaseSQLiteDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    $with(alias) {
        return {
            as(qb) {
                if (typeof qb === 'function') {
                    qb = qb(query_builders_1.queryBuilder);
                }
                return new Proxy(new subquery_1.WithSubquery(qb.getSQL(), qb.getSelectedFields(), alias, true), new subquery_1.SelectionProxyHandler({ alias, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
            },
        };
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new query_builders_1.SQLiteSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, self.session, self.dialect, queries);
        }
        return { select };
    }
    select(fields) {
        return new query_builders_1.SQLiteSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.SQLiteUpdateBuilder(table, this.session, this.dialect);
    }
    insert(into) {
        return new query_builders_1.SQLiteInsertBuilder(into, this.session, this.dialect);
    }
    delete(from) {
        return new query_builders_1.SQLiteDelete(from, this.session, this.dialect);
    }
    run(query) {
        return this.session.run(query.getSQL());
    }
    all(query) {
        return this.session.all(query.getSQL());
    }
    get(query) {
        return this.session.get(query.getSQL());
    }
    values(query) {
        return this.session.values(query.getSQL());
    }
    transaction(transaction, config) {
        return this.session.transaction(transaction, config);
    }
}
exports.BaseSQLiteDatabase = BaseSQLiteDatabase;
class SQLiteTransaction extends BaseSQLiteDatabase {
    constructor(dialect, session, nestedIndex = 0) {
        super(dialect, session);
        this.nestedIndex = nestedIndex;
    }
    rollback() {
        throw new errors_1.TransactionRollbackError();
    }
}
exports.SQLiteTransaction = SQLiteTransaction;
