// drizzle-orm/sqlite-core/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteSession = exports.PreparedQuery = void 0;
const errors_1 = require("../errors");
class PreparedQuery {
}
exports.PreparedQuery = PreparedQuery;
class SQLiteSession {
    constructor(
    /** @internal */
    dialect) {
        this.dialect = dialect;
    }
    prepareOneTimeQuery(query, fields) {
        return this.prepareQuery(query, fields);
    }
    run(query) {
        const staticQuery = this.dialect.sqlToQuery(query);
        try {
            return this.prepareOneTimeQuery(staticQuery, undefined).run();
        }
        catch (err) {
            throw errors_1.DrizzleError.wrap(err, `Failed to run the query '${staticQuery.sql}'`);
        }
    }
    all(query) {
        return this.prepareOneTimeQuery(this.dialect.sqlToQuery(query), undefined).all();
    }
    get(query) {
        return this.prepareOneTimeQuery(this.dialect.sqlToQuery(query), undefined).get();
    }
    values(query) {
        return this.prepareOneTimeQuery(this.dialect.sqlToQuery(query), undefined).values();
    }
}
exports.SQLiteSession = SQLiteSession;
