// drizzle-orm/sqlite-core/table.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.sqliteTableCreator = exports.sqliteTable = exports.SQLiteTable = exports.ExtraConfigBuilder = exports.InlineForeignKeys = void 0;
const table_1 = require("../table");
/** @internal */
exports.InlineForeignKeys = Symbol('InlineForeignKeys');
/** @internal */
exports.ExtraConfigBuilder = Symbol('ExtraConfigBuilder');
class SQLiteTable extends table_1.Table {
    constructor() {
        super(...arguments);
        /** @internal */
        this[_a] = [];
        /** @internal */
        this[_b] = undefined;
    }
}
table_1.Table.Symbol.Columns, _a = exports.InlineForeignKeys, _b = exports.ExtraConfigBuilder;
/** @internal */
SQLiteTable.Symbol = Object.assign({}, table_1.Table.Symbol, {
    InlineForeignKeys: exports.InlineForeignKeys,
    ExtraConfigBuilder: exports.ExtraConfigBuilder,
});
exports.SQLiteTable = SQLiteTable;
function sqliteTableBase(name, columns, extraConfig, schema, baseName = name) {
    const rawTable = new SQLiteTable(name, schema, baseName);
    const builtColumns = Object.fromEntries(Object.entries(columns).map(([name, colBuilder]) => {
        const column = colBuilder.build(rawTable);
        rawTable[exports.InlineForeignKeys].push(...colBuilder.buildForeignKeys(column, rawTable));
        return [name, column];
    }));
    const table = Object.assign(rawTable, builtColumns);
    table[table_1.Table.Symbol.Columns] = builtColumns;
    if (extraConfig) {
        table[SQLiteTable.Symbol.ExtraConfigBuilder] = extraConfig;
    }
    return table;
}
const sqliteTable = (name, columns, extraConfig) => {
    return sqliteTableBase(name, columns, extraConfig);
};
exports.sqliteTable = sqliteTable;
function sqliteTableCreator(customizeTableName) {
    return (name, columns, extraConfig) => {
        return sqliteTableBase(customizeTableName(name), columns, extraConfig, undefined, name);
    };
}
exports.sqliteTableCreator = sqliteTableCreator;
