// drizzle-orm/sqlite-core/view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.view = exports.sqliteView = exports.SQLiteView = exports.SQLiteViewConfig = exports.SQLiteViewBase = exports.ManualViewBuilder = exports.ViewBuilder = exports.ViewBuilderCore = void 0;
const subquery_1 = require("../subquery");
const utils_1 = require("../utils");
const view_1 = require("../view");
const query_builders_1 = require("./query-builders");
const table_1 = require("./table");
class ViewBuilderCore {
    constructor(name) {
        this.name = name;
        this.config = {};
    }
}
exports.ViewBuilderCore = ViewBuilderCore;
class ViewBuilder extends ViewBuilderCore {
    as(qb) {
        if (typeof qb === 'function') {
            qb = qb(query_builders_1.queryBuilder);
        }
        const selectionProxy = new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        });
        // const aliasedSelectedFields = new Proxy(qb.getSelectedFields(), selectionProxy);
        const aliasedSelectedFields = qb.getSelectedFields();
        return new Proxy(new SQLiteView({
            sqliteConfig: this.config,
            config: {
                name: this.name,
                schema: undefined,
                selectedFields: aliasedSelectedFields,
                query: qb.getSQL().inlineParams(),
            },
        }), selectionProxy);
    }
}
exports.ViewBuilder = ViewBuilder;
class ManualViewBuilder extends ViewBuilderCore {
    constructor(name, columns) {
        super(name);
        this.columns = (0, utils_1.getTableColumns)((0, table_1.sqliteTable)(name, columns));
    }
    existing() {
        return new Proxy(new SQLiteView({
            sqliteConfig: undefined,
            config: {
                name: this.name,
                schema: undefined,
                selectedFields: this.columns,
                query: undefined,
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
    as(query) {
        return new Proxy(new SQLiteView({
            sqliteConfig: this.config,
            config: {
                name: this.name,
                schema: undefined,
                selectedFields: this.columns,
                query: query.inlineParams(),
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
}
exports.ManualViewBuilder = ManualViewBuilder;
class SQLiteViewBase extends view_1.View {
}
exports.SQLiteViewBase = SQLiteViewBase;
exports.SQLiteViewConfig = Symbol('SQLiteViewConfig');
class SQLiteView extends SQLiteViewBase {
    constructor({ sqliteConfig, config }) {
        super(config);
        this[exports.SQLiteViewConfig] = sqliteConfig;
    }
}
exports.SQLiteView = SQLiteView;
function sqliteView(name, selection) {
    if (selection) {
        return new ManualViewBuilder(name, selection);
    }
    return new ViewBuilder(name);
}
exports.sqliteView = sqliteView;
exports.view = sqliteView;
