// drizzle-orm/node-postgres/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.NodePgDriver = void 0;
const pg_1 = require("pg");
const logger_1 = require("../logger");
const db_1 = require("../pg-core/db");
const dialect_1 = require("../pg-core/dialect");
const session_1 = require("./session");
class NodePgDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession() {
        return new session_1.NodePgSession(this.client, this.dialect, { logger: this.options.logger });
    }
    initMappers() {
        pg_1.types.setTypeParser(pg_1.types.builtins.TIMESTAMPTZ, (val) => val);
        pg_1.types.setTypeParser(pg_1.types.builtins.TIMESTAMP, (val) => val);
        pg_1.types.setTypeParser(pg_1.types.builtins.DATE, (val) => val);
    }
}
exports.NodePgDriver = NodePgDriver;
function drizzle(client, config = {}) {
    const dialect = new dialect_1.PgDialect();
    let logger;
    if (config.logger === true) {
        logger = new logger_1.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    const driver = new NodePgDriver(client, dialect, { logger });
    const session = driver.createSession();
    return new db_1.PgDatabase(dialect, session);
}
exports.drizzle = drizzle;
