// drizzle-orm/node-postgres/session.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodePgTransaction = exports.NodePgSession = exports.NodePgPreparedQuery = void 0;
const pg_1 = require("pg");
const logger_1 = require("../logger");
const pg_core_1 = require("../pg-core");
const session_1 = require("../pg-core/session");
const sql_1 = require("../sql");
const utils_1 = require("../utils");
class NodePgPreparedQuery extends session_1.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name) {
        super();
        this.client = client;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.rawQuery = {
            name,
            text: queryString,
        };
        this.query = {
            name,
            text: queryString,
            rowMode: 'array',
        };
    }
    execute(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        const { fields } = this;
        if (!fields) {
            return this.client.query(this.rawQuery, params);
        }
        const result = this.client.query(this.query, params);
        return result.then((result) => result.rows.map((row) => (0, utils_1.mapResultRow)(fields, row, this.joinsNotNullableMap)));
    }
    all(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.rawQuery, params).then((result) => result.rows);
    }
    values(placeholderValues = {}) {
        const params = (0, sql_1.fillPlaceholders)(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.query, params).then((result) => result.rows);
    }
}
exports.NodePgPreparedQuery = NodePgPreparedQuery;
class NodePgSession extends session_1.PgSession {
    constructor(client, dialect, options = {}) {
        var _a;
        super(dialect);
        this.client = client;
        this.options = options;
        this.logger = (_a = options.logger) !== null && _a !== void 0 ? _a : new logger_1.NoopLogger();
    }
    prepareQuery(query, fields, name) {
        return new NodePgPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name);
    }
    query(query, params) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logQuery(query, params);
            const result = yield this.client.query({
                rowMode: 'array',
                text: query,
                values: params,
            });
            return result;
        });
    }
    queryObjects(query, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.query(query, params);
        });
    }
    transaction(transaction, config) {
        return __awaiter(this, void 0, void 0, function* () {
            const session = this.client instanceof pg_1.Pool
                ? new NodePgSession(yield this.client.connect(), this.dialect, this.options)
                : this;
            const tx = new NodePgTransaction(this.dialect, session);
            yield tx.execute((0, sql_1.sql) `begin${config ? (0, sql_1.sql) ` ${tx.getTransactionConfigSQL(config)}` : undefined}`);
            try {
                const result = yield transaction(tx);
                yield tx.execute((0, sql_1.sql) `commit`);
                return result;
            }
            catch (error) {
                yield tx.execute((0, sql_1.sql) `rollback`);
                throw error;
            }
        });
    }
}
exports.NodePgSession = NodePgSession;
class NodePgTransaction extends pg_core_1.PgTransaction {
    transaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            const savepointName = `sp${this.nestedIndex + 1}`;
            const tx = new NodePgTransaction(this.dialect, this.session, this.nestedIndex + 1);
            yield tx.execute(sql_1.sql.raw(`savepoint ${savepointName}`));
            try {
                const result = yield transaction(tx);
                yield tx.execute(sql_1.sql.raw(`release savepoint ${savepointName}`));
                return result;
            }
            catch (err) {
                yield tx.execute(sql_1.sql.raw(`rollback to savepoint ${savepointName}`));
                throw err;
            }
        });
    }
}
exports.NodePgTransaction = NodePgTransaction;
