// drizzle-orm/mysql-core/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySqlDatabase = void 0;
const subquery_1 = require("../subquery");
const query_builders_1 = require("./query-builders");
class MySqlDatabase {
    constructor(
    /** @internal */
    dialect, 
    /** @internal */
    session) {
        this.dialect = dialect;
        this.session = session;
    }
    $with(alias) {
        return {
            as(qb) {
                if (typeof qb === 'function') {
                    qb = qb(query_builders_1.queryBuilder);
                }
                return new Proxy(new subquery_1.WithSubquery(qb.getSQL(), qb.getSelectedFields(), alias, true), new subquery_1.SelectionProxyHandler({ alias, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
            },
        };
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new query_builders_1.MySqlSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, self.session, self.dialect, queries);
        }
        return { select };
    }
    select(fields) {
        return new query_builders_1.MySqlSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, this.session, this.dialect);
    }
    update(table) {
        return new query_builders_1.MySqlUpdateBuilder(table, this.session, this.dialect);
    }
    insert(table) {
        return new query_builders_1.MySqlInsertBuilder(table, this.session, this.dialect);
    }
    delete(table) {
        return new query_builders_1.MySqlDelete(table, this.session, this.dialect);
    }
    execute(query) {
        return this.session.execute(query.getSQL());
    }
    transaction(transaction, config) {
        return this.session.transaction(transaction, config);
    }
}
exports.MySqlDatabase = MySqlDatabase;
