// drizzle-orm/mysql-core/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mysqlSchema = exports.mysqlDatabase = exports.isMySqlSchema = exports.MySqlSchema = void 0;
const table_1 = require("./table");
const view_1 = require("./view");
class MySqlSchema {
    constructor(schemaName) {
        this.schemaName = schemaName;
        this.table = (name, columns, extraConfig) => {
            return (0, table_1.mysqlTableWithSchema)(name, columns, extraConfig, this.schemaName);
        };
        this.view = ((name, columns) => {
            return (0, view_1.mysqlViewWithSchema)(name, columns, this.schemaName);
        });
    }
}
exports.MySqlSchema = MySqlSchema;
/** @deprecated - use `instanceof MySqlSchema` */
function isMySqlSchema(obj) {
    return obj instanceof MySqlSchema;
}
exports.isMySqlSchema = isMySqlSchema;
/**
 * Create a MySQL schema.
 * https://dev.mysql.com/doc/refman/8.0/en/create-database.html
 *
 * @param name mysql use schema name
 * @returns MySQL schema
 */
function mysqlDatabase(name) {
    return new MySqlSchema(name);
}
exports.mysqlDatabase = mysqlDatabase;
/**
 * @see mysqlDatabase
 */
exports.mysqlSchema = mysqlDatabase;
