// drizzle-orm/mysql-core/table.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.mysqlTableCreator = exports.mysqlTable = exports.mysqlTableWithSchema = exports.MySqlTable = exports.ExtraConfigBuilder = exports.InlineForeignKeys = void 0;
const table_1 = require("../table");
/** @internal */
exports.InlineForeignKeys = Symbol('InlineForeignKeys');
/** @internal */
exports.ExtraConfigBuilder = Symbol('ExtraConfigBuilder');
class MySqlTable extends table_1.Table {
    constructor() {
        super(...arguments);
        /** @internal */
        this[_a] = [];
        /** @internal */
        this[_b] = undefined;
    }
}
table_1.Table.Symbol.Columns, _a = exports.InlineForeignKeys, _b = exports.ExtraConfigBuilder;
/** @internal */
MySqlTable.Symbol = Object.assign({}, table_1.Table.Symbol, {
    InlineForeignKeys: exports.InlineForeignKeys,
    ExtraConfigBuilder: exports.ExtraConfigBuilder,
});
exports.MySqlTable = MySqlTable;
function mysqlTableWithSchema(name, columns, extraConfig, schema, baseName = name) {
    const rawTable = new MySqlTable(name, schema, baseName);
    const builtColumns = Object.fromEntries(Object.entries(columns).map(([name, colBuilder]) => {
        const column = colBuilder.build(rawTable);
        rawTable[exports.InlineForeignKeys].push(...colBuilder.buildForeignKeys(column, rawTable));
        return [name, column];
    }));
    const table = Object.assign(rawTable, builtColumns);
    table[table_1.Table.Symbol.Columns] = builtColumns;
    if (extraConfig) {
        table[MySqlTable.Symbol.ExtraConfigBuilder] = extraConfig;
    }
    return table;
}
exports.mysqlTableWithSchema = mysqlTableWithSchema;
const mysqlTable = (name, columns, extraConfig) => {
    return mysqlTableWithSchema(name, columns, extraConfig, undefined, name);
};
exports.mysqlTable = mysqlTable;
function mysqlTableCreator(customizeTableName) {
    return (name, columns, extraConfig) => {
        return mysqlTableWithSchema(customizeTableName(name), columns, extraConfig, undefined, name);
    };
}
exports.mysqlTableCreator = mysqlTableCreator;
