// drizzle-orm/mysql-core/view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mysqlView = exports.mysqlViewWithSchema = exports.MySqlView = exports.MySqlViewConfig = exports.MySqlViewBase = exports.ManualViewBuilder = exports.ViewBuilder = exports.ViewBuilderCore = void 0;
const subquery_1 = require("../subquery");
const utils_1 = require("../utils");
const view_1 = require("../view");
const query_builders_1 = require("./query-builders");
const table_1 = require("./table");
class ViewBuilderCore {
    constructor(name, schema) {
        this.name = name;
        this.schema = schema;
        this.config = {};
    }
    algorithm(algorithm) {
        this.config.algorithm = algorithm;
        return this;
    }
    definer(definer) {
        this.config.definer = definer;
        return this;
    }
    sqlSecurity(sqlSecurity) {
        this.config.sqlSecurity = sqlSecurity;
        return this;
    }
    withCheckOption(withCheckOption) {
        this.config.withCheckOption = withCheckOption !== null && withCheckOption !== void 0 ? withCheckOption : 'cascaded';
        return this;
    }
}
exports.ViewBuilderCore = ViewBuilderCore;
class ViewBuilder extends ViewBuilderCore {
    as(qb) {
        if (typeof qb === 'function') {
            qb = qb(query_builders_1.queryBuilder);
        }
        const selectionProxy = new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        });
        const aliasedSelection = new Proxy(qb.getSelectedFields(), selectionProxy);
        return new Proxy(new MySqlView({
            mysqlConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: aliasedSelection,
                query: qb.getSQL().inlineParams(),
            },
        }), selectionProxy);
    }
}
exports.ViewBuilder = ViewBuilder;
class ManualViewBuilder extends ViewBuilderCore {
    constructor(name, columns, schema) {
        super(name, schema);
        this.columns = (0, utils_1.getTableColumns)((0, table_1.mysqlTable)(name, columns));
    }
    existing() {
        return new Proxy(new MySqlView({
            mysqlConfig: undefined,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: undefined,
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
    as(query) {
        return new Proxy(new MySqlView({
            mysqlConfig: this.config,
            config: {
                name: this.name,
                schema: this.schema,
                selectedFields: this.columns,
                query: query.inlineParams(),
            },
        }), new subquery_1.SelectionProxyHandler({
            alias: this.name,
            sqlBehavior: 'error',
            sqlAliasedBehavior: 'alias',
            replaceOriginalName: true,
        }));
    }
}
exports.ManualViewBuilder = ManualViewBuilder;
class MySqlViewBase extends view_1.View {
}
exports.MySqlViewBase = MySqlViewBase;
exports.MySqlViewConfig = Symbol('MySqlViewConfig');
class MySqlView extends MySqlViewBase {
    constructor({ mysqlConfig, config }) {
        super(config);
        this[exports.MySqlViewConfig] = mysqlConfig;
    }
}
exports.MySqlView = MySqlView;
/** @internal */
function mysqlViewWithSchema(name, selection, schema) {
    if (selection) {
        return new ManualViewBuilder(name, selection, schema);
    }
    return new ViewBuilder(name, schema);
}
exports.mysqlViewWithSchema = mysqlViewWithSchema;
function mysqlView(name, selection) {
    return mysqlViewWithSchema(name, selection, undefined);
}
exports.mysqlView = mysqlView;
