// drizzle-orm/sqlite-core/query-builders/query-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryBuilder = exports.QueryBuilderInstance = void 0;
const dialect_1 = require("../dialect");
const subquery_1 = require("../../subquery");
class QueryBuilderInstance {
    constructor() {
        // Required to avoid circular dependency
        this.SQLiteSelectBuilder = require('./select').SQLiteSelectBuilder;
    }
    $with(alias) {
        const queryBuilder = this;
        return {
            as(qb) {
                if (typeof qb === 'function') {
                    qb = qb(queryBuilder);
                }
                return new Proxy(new subquery_1.WithSubquery(qb.getSQL(), qb.getSelectedFields(), alias, true), new subquery_1.SelectionProxyHandler({ alias, sqlAliasedBehavior: 'alias', sqlBehavior: 'error' }));
            },
        };
    }
    with(...queries) {
        const self = this;
        function select(fields) {
            return new self.SQLiteSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, undefined, self.getDialect(), queries);
        }
        return { select };
    }
    select(fields) {
        return new this.SQLiteSelectBuilder(fields !== null && fields !== void 0 ? fields : undefined, undefined, this.getDialect());
    }
    // Lazy load dialect to avoid circular dependency
    getDialect() {
        if (!this.dialect) {
            this.dialect = new dialect_1.SQLiteSyncDialect();
        }
        return this.dialect;
    }
}
exports.QueryBuilderInstance = QueryBuilderInstance;
exports.queryBuilder = new QueryBuilderInstance();
