// drizzle-orm/sqlite-core/query-builders/update.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteUpdate = exports.SQLiteUpdateBuilder = void 0;
const table_1 = require("../table");
const utils_1 = require("../../utils");
class SQLiteUpdateBuilder {
    constructor(table, session, dialect) {
        this.table = table;
        this.session = session;
        this.dialect = dialect;
    }
    set(values) {
        return new SQLiteUpdate(this.table, (0, utils_1.mapUpdateSet)(this.table, values), this.session, this.dialect);
    }
}
exports.SQLiteUpdateBuilder = SQLiteUpdateBuilder;
class SQLiteUpdate {
    constructor(table, set, session, dialect) {
        this.session = session;
        this.dialect = dialect;
        this.run = (placeholderValues) => {
            return this.prepare(true).run(placeholderValues);
        };
        this.all = (placeholderValues) => {
            return this.prepare(true).all(placeholderValues);
        };
        this.get = (placeholderValues) => {
            return this.prepare(true).get(placeholderValues);
        };
        this.values = (placeholderValues) => {
            return this.prepare(true).values(placeholderValues);
        };
        this.config = { set, table };
    }
    where(where) {
        this.config.where = where;
        return this;
    }
    returning(fields = this.config.table[table_1.SQLiteTable.Symbol.Columns]) {
        this.config.returning = (0, utils_1.orderSelectedFields)(fields);
        return this;
    }
    /** @internal */
    getSQL() {
        return this.dialect.buildUpdateQuery(this.config);
    }
    toSQL() {
        const _a = this.dialect.sqlToQuery(this.getSQL()), { typings } = _a, rest = __rest(_a, ["typings"]);
        return rest;
    }
    prepare(isOneTimeQuery) {
        return this.session[isOneTimeQuery ? 'prepareOneTimeQuery' : 'prepareQuery'](this.dialect.sqlToQuery(this.getSQL()), this.config.returning);
    }
}
exports.SQLiteUpdate = SQLiteUpdate;
