// drizzle-orm/aws-data-api/pg/driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.drizzle = exports.AwsPgDialect = exports.AwsDataApiDriver = void 0;
const logger_1 = require("../../logger");
const db_1 = require("../../pg-core/db");
const dialect_1 = require("../../pg-core/dialect");
const session_1 = require("./session");
class AwsDataApiDriver {
    constructor(client, dialect, options) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
    }
    createSession() {
        return new session_1.AwsDataApiSession(this.client, this.dialect, this.options, undefined);
    }
}
exports.AwsDataApiDriver = AwsDataApiDriver;
class AwsPgDialect extends dialect_1.PgDialect {
    escapeName(name) {
        return `"${name}"`;
    }
    escapeParam(num) {
        return `:${num + 1}`;
    }
}
exports.AwsPgDialect = AwsPgDialect;
function drizzle(client, config) {
    const dialect = new AwsPgDialect();
    let logger;
    if (config.logger === true) {
        logger = new logger_1.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    const driver = new AwsDataApiDriver(client, dialect, Object.assign(Object.assign({}, config), { logger }));
    const session = driver.createSession();
    return new db_1.PgDatabase(dialect, session);
}
exports.drizzle = drizzle;
