// drizzle-orm/sqlite-core/columns/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customType = exports.SQLiteCustomColumn = exports.SQLiteCustomColumnBuilder = void 0;
const common_1 = require("./common");
class SQLiteCustomColumnBuilder extends common_1.SQLiteColumnBuilder {
    constructor(name, fieldConfig, customTypeParams) {
        super(name);
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new SQLiteCustomColumn(table, this.config);
    }
}
exports.SQLiteCustomColumnBuilder = SQLiteCustomColumnBuilder;
class SQLiteCustomColumn extends common_1.SQLiteColumn {
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        if (typeof this.mapFrom === 'function') {
            return this.mapFrom(value);
        }
        else {
            return value;
        }
    }
    mapToDriverValue(value) {
        if (typeof this.mapTo === 'function') {
            return this.mapTo(value);
        }
        else {
            return value;
        }
    }
}
exports.SQLiteCustomColumn = SQLiteCustomColumn;
/**
 * Custom sqlite database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new SQLiteCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}
exports.customType = customType;
