// drizzle-orm/sqlite-core/columns/integer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.int = exports.integer = exports.SQLiteTimestamp = exports.SQLiteTimestampBuilder = exports.SQLiteInteger = exports.SQLiteIntegerBuilder = exports.SQLiteBaseInteger = exports.SQLiteBaseIntegerBuilder = void 0;
const sql_1 = require("../../sql");
const common_1 = require("./common");
class SQLiteBaseIntegerBuilder extends common_1.SQLiteColumnBuilder {
    constructor(name) {
        super(name);
        this.config.autoIncrement = false;
    }
    primaryKey(config) {
        if (config === null || config === void 0 ? void 0 : config.autoIncrement) {
            this.config.autoIncrement = true;
        }
        this.config.hasDefault = true;
        return super.primaryKey();
    }
}
exports.SQLiteBaseIntegerBuilder = SQLiteBaseIntegerBuilder;
class SQLiteBaseInteger extends common_1.SQLiteColumn {
    constructor() {
        super(...arguments);
        this.autoIncrement = this.config.autoIncrement;
    }
    getSQLType() {
        return 'integer';
    }
}
exports.SQLiteBaseInteger = SQLiteBaseInteger;
class SQLiteIntegerBuilder extends SQLiteBaseIntegerBuilder {
    build(table) {
        return new SQLiteInteger(table, this.config);
    }
}
exports.SQLiteIntegerBuilder = SQLiteIntegerBuilder;
class SQLiteInteger extends SQLiteBaseInteger {
}
exports.SQLiteInteger = SQLiteInteger;
class SQLiteTimestampBuilder extends SQLiteBaseIntegerBuilder {
    constructor(name, mode) {
        super(name);
        this.config.mode = mode;
    }
    /**
     * @deprecated Use `default()` with your own expression instead.
     *
     * Adds `DEFAULT (cast((julianday('now') - 2440587.5)*86400000 as integer))` to the column, which is the current epoch timestamp in milliseconds.
     */
    defaultNow() {
        return this.default((0, sql_1.sql) `(cast((julianday('now') - 2440587.5)*86400000 as integer))`);
    }
    build(table) {
        return new SQLiteTimestamp(table, this.config);
    }
}
exports.SQLiteTimestampBuilder = SQLiteTimestampBuilder;
class SQLiteTimestamp extends SQLiteBaseInteger {
    constructor() {
        super(...arguments);
        this.mode = this.config.mode;
    }
    mapFromDriverValue(value) {
        if (this.config.mode === 'timestamp') {
            return new Date(value * 1000);
        }
        return new Date(value);
    }
    mapToDriverValue(value) {
        const unix = value.getTime();
        if (this.config.mode === 'timestamp') {
            return Math.floor(unix / 1000);
        }
        return unix;
    }
}
exports.SQLiteTimestamp = SQLiteTimestamp;
function integer(name, config) {
    if ((config === null || config === void 0 ? void 0 : config.mode) === 'timestamp' || (config === null || config === void 0 ? void 0 : config.mode) === 'timestamp_ms') {
        return new SQLiteTimestampBuilder(name, config.mode);
    }
    return new SQLiteIntegerBuilder(name);
}
exports.integer = integer;
exports.int = integer;
