// drizzle-orm/sql/expressions/conditions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.notIlike = exports.ilike = exports.notLike = exports.like = exports.notBetween = exports.between = exports.notExists = exports.exists = exports.isNotNull = exports.isNull = exports.notInArray = exports.inArray = exports.lte = exports.lt = exports.gte = exports.gt = exports.not = exports.or = exports.and = exports.ne = exports.eq = exports.bindIfParam = void 0;
const column_1 = require("../../column");
const table_1 = require("../../table");
const view_1 = require("../../view");
const index_1 = require("../index");
function bindIfParam(value, column) {
    if ((0, index_1.isDriverValueEncoder)(column) && !(0, index_1.isSQLWrapper)(value) && !(value instanceof index_1.Param) && !(value instanceof index_1.Placeholder)
        && !(value instanceof column_1.Column) && !(value instanceof table_1.Table) && !(value instanceof view_1.View)) {
        return new index_1.Param(value, column);
    }
    return value;
}
exports.bindIfParam = bindIfParam;
function eq(left, right) {
    return (0, index_1.sql) `${left} = ${bindIfParam(right, left)}`;
}
exports.eq = eq;
function ne(left, right) {
    return (0, index_1.sql) `${left} <> ${bindIfParam(right, left)}`;
}
exports.ne = ne;
function and(...conditions) {
    if (conditions.length === 0) {
        return undefined;
    }
    const chunks = [index_1.sql.raw('(')];
    conditions
        .filter((c) => typeof c !== 'undefined')
        .forEach((condition, index) => {
        if (index === 0) {
            chunks.push(condition);
        }
        else {
            chunks.push((0, index_1.sql) ` and `, condition);
        }
    });
    chunks.push((0, index_1.sql) `)`);
    return index_1.sql.fromList(chunks);
}
exports.and = and;
function or(...conditions) {
    if (conditions.length === 0) {
        return undefined;
    }
    const chunks = [index_1.sql.raw('(')];
    conditions
        .filter((c) => typeof c !== 'undefined')
        .forEach((condition, index) => {
        if (index === 0) {
            chunks.push(condition);
        }
        else {
            chunks.push((0, index_1.sql) ` or `, condition);
        }
    });
    chunks.push((0, index_1.sql) `)`);
    return index_1.sql.fromList(chunks);
}
exports.or = or;
function not(condition) {
    return (0, index_1.sql) `not ${condition}`;
}
exports.not = not;
function gt(left, right) {
    return (0, index_1.sql) `${left} > ${bindIfParam(right, left)}`;
}
exports.gt = gt;
function gte(left, right) {
    return (0, index_1.sql) `${left} >= ${bindIfParam(right, left)}`;
}
exports.gte = gte;
function lt(left, right) {
    return (0, index_1.sql) `${left} < ${bindIfParam(right, left)}`;
}
exports.lt = lt;
function lte(left, right) {
    return (0, index_1.sql) `${left} <= ${bindIfParam(right, left)}`;
}
exports.lte = lte;
function inArray(column, values) {
    if (Array.isArray(values)) {
        if (values.length === 0) {
            throw new Error('inArray requires at least one value');
        }
        return (0, index_1.sql) `${column} in ${values.map((v) => bindIfParam(v, column))}`;
    }
    return (0, index_1.sql) `${column} in ${bindIfParam(values, column)}`;
}
exports.inArray = inArray;
function notInArray(column, values) {
    if ((0, index_1.isSQLWrapper)(values)) {
        return (0, index_1.sql) `${column} not in ${values}`;
    }
    if (Array.isArray(values)) {
        if (values.length === 0) {
            throw new Error('inArray requires at least one value');
        }
        return (0, index_1.sql) `${column} not in ${values.map((v) => bindIfParam(v, column))}`;
    }
    return (0, index_1.sql) `${column} not in ${bindIfParam(values, column)}`;
}
exports.notInArray = notInArray;
function isNull(column) {
    return (0, index_1.sql) `${column} is null`;
}
exports.isNull = isNull;
function isNotNull(column) {
    return (0, index_1.sql) `${column} is not null`;
}
exports.isNotNull = isNotNull;
function exists(subquery) {
    return (0, index_1.sql) `exists (${subquery})`;
}
exports.exists = exists;
function notExists(subquery) {
    return (0, index_1.sql) `exists (${subquery})`;
}
exports.notExists = notExists;
function between(column, min, max) {
    return (0, index_1.sql) `${column} between ${bindIfParam(min, column)} and ${bindIfParam(max, column)}`;
}
exports.between = between;
function notBetween(column, min, max) {
    return (0, index_1.sql) `${column} not between ${bindIfParam(min, column)} and ${bindIfParam(max, column)}`;
}
exports.notBetween = notBetween;
function like(column, value) {
    return (0, index_1.sql) `${column} like ${value}`;
}
exports.like = like;
function notLike(column, value) {
    return (0, index_1.sql) `${column} not like ${value}`;
}
exports.notLike = notLike;
function ilike(column, value) {
    return (0, index_1.sql) `${column} ilike ${value}`;
}
exports.ilike = ilike;
function notIlike(column, value) {
    return (0, index_1.sql) `${column} not ilike ${value}`;
}
exports.notIlike = notIlike;
